diff --git a/runtime/caml/misc.h b/runtime/caml/misc.h
index c605f87..43e8762 100644
--- a/runtime/caml/misc.h
+++ b/runtime/caml/misc.h
@@ -116,6 +116,8 @@ CAMLdeprecated_typedef(addr, char *);
   #if defined(SUPPORT_DYNAMIC_LINKING) && defined(ARCH_SIXTYFOUR) \
       && (defined(__CYGWIN__) || defined(__MINGW32__))
     #define CAMLDLLIMPORT __declspec(dllimport)
+  #elif defined(SUPPORT_DYNAMIC_LINKING) && defined(_MSC_VER)
+    #define CAMLDLLIMPORT __declspec(dllimport)
   #else
     #define CAMLDLLIMPORT
   #endif
