diff --git a/Makefile b/Makefile
index 1add5117c7bfea540e4e10b782689fbd0e254dfe..bdddc37400cd2b6512e6b84720329b971b46581a 100644
--- a/Makefile
+++ b/Makefile
@@ -38,7 +38,7 @@ INCLUDES=-I utils -I parsing -I typing -I bytecomp -I file_formats \
         -I lambda -I middle_end -I middle_end/closure \
         -I middle_end/flambda -I middle_end/flambda/base_types \
         -I asmcomp \
-        -I driver -I toplevel
+        -I driver -I toplevel -I runtime
 
 COMPFLAGS=-strict-sequence -principal -absname \
           -w +a-4-9-40-41-42-44-45-48-66-70 \
@@ -71,7 +71,7 @@ TOPLEVELINIT=toplevel/toploop.cmo
 # capitalized module names.
 PERVASIVES=$(STDLIB_MODULES) outcometree topdirs toploop
 
-LIBFILES=stdlib.cma std_exit.cmo *.cmi camlheader
+LIBFILES=stdlib.cma std_exit.cmo *.cmi $(HEADER_NAME)
 
 COMPLIBDIR=$(LIBDIR)/compiler-libs
 
@@ -1151,7 +1151,7 @@ distclean: clean
 	$(MAKE) -C ocamltest distclean
 	$(MAKE) -C runtime distclean
 	$(MAKE) -C stdlib distclean
-	rm -f boot/ocamlrun boot/ocamlrun.exe boot/camlheader \
+	rm -f boot/ocamlrun boot/ocamlrun.exe boot/$(HEADER_NAME) \
 	      boot/ocamlruns boot/ocamlruns.exe \
 	      boot/flexlink.byte boot/flexlink.byte.exe \
 	      boot/flexdll_*.o boot/flexdll_*.obj \
diff --git a/Makefile.build_config.in b/Makefile.build_config.in
index 8f9b5c3863b3a0460fedc30ae548961c848540b0..f19f2625860adc2a9cd324c75eaa0d627aa54a91 100644
--- a/Makefile.build_config.in
+++ b/Makefile.build_config.in
@@ -52,5 +52,8 @@ PACKAGE_TARNAME = @PACKAGE_TARNAME@
 datarootdir = @datarootdir@
 DOCDIR=@docdir@
 
+### Where to look for the standard library on target
+TARGET_LIBDIR=@TARGET_LIBDIR@
+
 # Platform-dependent command to create symbolic links
 LN = @ln@
diff --git a/Makefile.common b/Makefile.common
index 60b9f66f8a44f5606ba46db118779d4b685ff8bb..93eeb9a5045060030b8c02a9d8753fadfd797d0f 100644
--- a/Makefile.common
+++ b/Makefile.common
@@ -173,3 +173,20 @@ OCAMLYACCFLAGS ?=
 	$(OCAMLYACC) $(OCAMLYACCFLAGS) $<
 
 SAK = $(ROOTDIR)/runtime/sak$(EXE)
+
+# ocamlc has several mechanisms for linking a bytecode image to the runtime
+# which executes it. The exact mechanism depends on the platform and the precise
+# location of the runtime. The boot compiler (boot/ocamlc) therefore needs this
+# additional information, which comes from the stdlib build and is copied into
+# boot/ as part of coldstart. See read_runtime_launch_info in
+# bytecomp/bytelink.ml for further details.
+HEADER_NAME = camlheader
+
+ifeq "$(UNIX_OR_WIN32)" "win32"
+# Ensure that no command can create Cygwin symbolic links by ensuring that
+# symlink(2) will fail if native NTFS symlinks aren't available.
+export CYGWIN := $(strip \
+  $(filter-out winsymlinks winsymlinks:%, $(CYGWIN)) winsymlinks:nativestrict)
+export MSYS := $(strip \
+  $(filter-out winsymlinks winsymlinks:%, $(MSYS)) winsymlinks:nativestrict)
+endif
diff --git a/bytecomp/bytelink.ml b/bytecomp/bytelink.ml
index ba6a3c39f51f75b6921635d5de5b4586fb65fde4..04d5bdb4729ae259dec62f5d2b8a2be3dfcbba51 100644
--- a/bytecomp/bytelink.ml
+++ b/bytecomp/bytelink.ml
@@ -304,6 +304,165 @@ let make_absolute file =
   else Location.rewrite_absolute_path
          (Filename.concat (Sys.getcwd()) file)
 
+type launch_method =
+| Shebang_bin_sh of string
+| Shebang_runtime
+| Executable
+
+type runtime_launch_info = {
+  buffer : string;
+  bindir : string;
+  launcher : launch_method;
+  executable_offset : int
+}
+
+(* See https://www.in-ulm.de/~mascheck/various/shebang/#origin for a deep
+   dive into shebangs.
+   - Whitespace (space or horizontal tab) delimits the interpreter from an
+     optional argument
+   - The path clearly must not contain a linefeed
+   - A maximum length of 125 (128 less the #! and the newline) is picked as a
+     portable maximum (it's actually Linux's prior to kernel v5.1), rather than
+     actually probing the maximum length in configure *)
+let invalid_for_shebang_line path =
+  let invalid_char = function ' ' | '\t' | '\n' -> true | _ -> false in
+  String.length path > 125 || String.exists invalid_char path
+
+(* The runtime-launch-info file consists of two "lines" followed by binary data.
+   The file is _always_ LF-formatted, even on Windows. The sequence of bytes up
+   to the first '\n' is interpreted:
+     - "sh" - use a shebang-style launcher. If sh is needed, determine its
+              location from [command -p -v sh]
+     - "exe" - use the executable launcher contained in this runtime-launch-info
+               file.
+     - "/" ^ path - use a shebang-style launcher. If sh is needed, path is the
+                    absolute location of sh. path must be valid for a shebang
+                    line.
+   The second "line" is interpreted as the next "\000\n"-terminated sequence and
+   is the directory containing the default runtimes (ocamlrun, ocamlrund, etc.).
+   The null terminator is used since '\n' is valid in a nefarious installation
+   prefix but Posix forbids filenames including the nul character.
+   The remainder of the file is then the executable launcher for bytecode
+   programs (see stdlib/header{,nt}.c). *)
+
+let read_runtime_launch_info file =
+  let buffer =
+    try
+      In_channel.with_open_bin file In_channel.input_all
+    with Sys_error msg -> raise (Error (Camlheader (msg, file)))
+  in
+  try
+    let bindir_start = String.index buffer '\n' + 1 in
+    let bindir_end = String.index_from buffer bindir_start '\000' in
+    let bindir = String.sub buffer bindir_start (bindir_end - bindir_start) in
+    let executable_offset = bindir_end + 2 in
+    let launcher =
+      let kind = String.sub buffer 0 (bindir_start - 1) in
+      if kind = "exe" then
+        Executable
+      else if kind <> "" && (kind.[0] = '/' || kind = "sh") then
+        Shebang_bin_sh kind
+      else
+        raise Not_found in
+    if String.length buffer < executable_offset
+       || buffer.[executable_offset - 1] <> '\n' then
+      raise Not_found
+    else
+      {bindir; launcher; buffer; executable_offset}
+  with Not_found ->
+    raise (Error (Camlheader ("corrupt header", file)))
+
+let find_bin_sh () =
+  let output_file = Filename.temp_file "caml_bin_sh" "" in
+  let result =
+  try
+    let cmd =
+      Filename.quote_command ~stdout:output_file "command" ["-p"; "-v"; "sh"]
+    in
+    if !Clflags.verbose then
+      Printf.eprintf "+ %s\n" cmd;
+    if Sys.command cmd = 0 then
+      In_channel.with_open_text output_file input_line
+    else
+      ""
+  with Sys_error _
+     | End_of_file -> ""
+  in
+  remove_file output_file;
+  result
+
+(* Writes the executable header to outchan and writes the RNTM section, if
+   needed. Returns a toc_writer (i.e. Bytesections.init_record is always
+   called) *)
+
+let write_header outchan =
+  let use_runtime, runtime =
+    if String.length !Clflags.use_runtime > 0 then
+      (true, make_absolute !Clflags.use_runtime)
+    else
+      (false, "ocamlrun" ^ !Clflags.runtime_variant)
+  in
+  (* Write the header *)
+  let runtime_info =
+    let header = "camlheader" in
+    try read_runtime_launch_info (Load_path.find header)
+    with Not_found -> raise (Error (File_not_found header))
+  in
+  let runtime =
+    (* Historically, the native Windows ports are assumed to be finding
+       ocamlrun using a PATH search. *)
+    if use_runtime || Sys.win32 then
+      runtime
+    else
+      Filename.concat runtime_info.bindir runtime
+  in
+  (* Determine which method will be used for launching the executable:
+     Executable: concatenate the bytecode image to the executable stub
+     Shebang_runtime: #! line with the required runtime
+     Shebang_bin_sh: #! for a shell script calling exec *)
+  let launcher =
+    if runtime_info.launcher = Executable then
+      Executable
+    else
+      if invalid_for_shebang_line runtime then
+        match runtime_info.launcher with
+        | Shebang_bin_sh sh ->
+            let sh =
+              if sh = "sh" then
+                find_bin_sh ()
+              else
+                sh in
+            if sh = "" || invalid_for_shebang_line sh then
+              Executable
+            else
+              Shebang_bin_sh sh
+        | _ ->
+            Executable
+      else
+        Shebang_runtime
+  in
+  match launcher with
+  | Shebang_runtime ->
+      (* Use the runtime directly *)
+      Printf.fprintf outchan "#!%s\n" runtime;
+      Bytesections.init_record outchan
+  | Shebang_bin_sh bin_sh ->
+      (* exec the runtime using sh *)
+      Printf.fprintf outchan "\
+        #!%s\n\
+        exec %s \"$0\" \"$@\"\n" bin_sh (Filename.quote runtime);
+      Bytesections.init_record outchan
+  | Executable ->
+      (* Use the executable stub launcher *)
+      let pos = runtime_info.executable_offset in
+      let len = String.length runtime_info.buffer - pos in
+      Out_channel.output_substring outchan runtime_info.buffer pos len;
+      (* The runtime name needs recording in RNTM *)
+      let toc_writer = Bytesections.init_record outchan in
+      Printf.fprintf outchan "%s\000" runtime;
+      Bytesections.record outchan "RNTM";
+      toc_writer
+
 (* Create a bytecode executable file *)
 
 let link_bytecode ?final_name tolink exec_name standalone =
@@ -323,37 +482,11 @@ let link_bytecode ?final_name tolink exec_name standalone =
     ~always:(fun () -> close_out outchan)
     ~exceptionally:(fun () -> remove_file exec_name)
     (fun () ->
-       if standalone && !Clflags.with_runtime then begin
-         (* Copy the header *)
-         let header =
-           if String.length !Clflags.use_runtime > 0
-           then "camlheader_ur" else "camlheader" ^ !Clflags.runtime_variant
-         in
-         try
-           let inchan = open_in_bin (Load_path.find header) in
-           copy_file inchan outchan;
-           close_in inchan
-         with
-         | Not_found -> raise (Error (File_not_found header))
-         | Sys_error msg -> raise (Error (Camlheader (header, msg)))
-       end;
-       Bytesections.init_record outchan;
-       (* The path to the bytecode interpreter (in use_runtime mode) *)
-       if String.length !Clflags.use_runtime > 0 && !Clflags.with_runtime then
-       begin
-         let runtime = make_absolute !Clflags.use_runtime in
-         let runtime =
-           (* shebang mustn't exceed 128 including the #! and \0 *)
-           if String.length runtime > 125 then
-             "/bin/sh\n\
-              exec \"" ^ runtime ^ "\" \"$0\" \"$@\""
-           else
-             runtime
-         in
-         output_string outchan runtime;
-         output_char outchan '\n';
-         Bytesections.record outchan "RNTM"
-       end;
+       (* Write the header and set the path to the bytecode interpreter *)
+       if standalone && !Clflags.with_runtime then
+         write_header outchan
+       else
+         Bytesections.init_record outchan;
        (* The bytecode *)
        let start_code = pos_out outchan in
        Symtable.init();
@@ -472,6 +605,13 @@ let guarded_primitives = [
     "caml_set_oo_id", "caml__set_oo_id";
   ]
 
+let output_without_guarded_primitives outchan s =
+  List.iter (fun (f, f') -> Printf.fprintf outchan "\n#define %s %s" f f')
+    guarded_primitives;
+  output_string outchan s;
+  List.iter (fun (f, _) -> Printf.fprintf outchan "\n#undef %s" f)
+    guarded_primitives
+
 let link_bytecode_as_c tolink outfile with_main =
   let outchan = open_out outfile in
   Misc.try_finally
@@ -486,15 +626,11 @@ let link_bytecode_as_c tolink outfile with_main =
 \n#ifdef __cplusplus\
 \nextern \"C\" {\
 \n#endif";
-       List.iter (fun (f, f') -> Printf.fprintf outchan "\n#define %s %s" f f')
-         guarded_primitives;
-       output_string outchan "\
+       output_without_guarded_primitives outchan "\
 \n#include <caml/mlvalues.h>\
 \n#include <caml/startup.h>\
 \n#include <caml/sys.h>\
 \n#include <caml/misc.h>\n";
-       List.iter (fun (f, _) -> Printf.fprintf outchan "\n#undef %s" f)
-         guarded_primitives;
        output_string outchan "\nstatic int caml_code[] = {\n";
        Symtable.init();
        clear_crc_interfaces ();
@@ -589,23 +725,21 @@ let build_custom_runtime prim_name exec_name =
     if not !Clflags.with_runtime
     then ""
     else "-lcamlrun" ^ !Clflags.runtime_variant in
-  let debug_prefix_map =
-    if Config.c_has_debug_prefix_map && not !Clflags.keep_camlprimc_file then
-      let flag =
-        [Printf.sprintf "-fdebug-prefix-map=%s=camlprim.c" prim_name]
-      in
-        if Ccomp.linker_is_flexlink then
-          "-link" :: flag
-        else
-          flag
+  let stable_name =
+    if not !Clflags.keep_camlprimc_file then
+      Some "camlprim.c"
     else
-      [] in
-  let exitcode =
-    (Clflags.std_include_flag "-I" ^ " " ^ Config.bytecomp_c_libraries)
+      None
+  in
+  let prims_obj = Filename.temp_file "camlprim" Config.ext_obj in
+  let result =
+    Ccomp.compile_file ~output:prims_obj ?stable_name prim_name = 0
+    && Ccomp.call_linker Ccomp.Exe exec_name
+        ([prims_obj] @ List.rev !Clflags.ccobjs @ [runtime_lib])
+        (Clflags.std_include_flag "-I" ^ " " ^ Config.bytecomp_c_libraries) = 0
   in
-  Ccomp.call_linker Ccomp.Exe exec_name
-    (debug_prefix_map @ [prim_name] @ List.rev !Clflags.ccobjs @ [runtime_lib])
-    exitcode = 0
+  remove_file prims_obj;
+  result
 
 let append_bytecode bytecode_name exec_name =
   let oc = open_out_gen [Open_wronly; Open_append; Open_binary] 0 exec_name in
@@ -675,16 +809,10 @@ let link objfiles output_name =
          output_string poc "\
          #ifdef __cplusplus\n\
          extern \"C\" {\n\
-         #endif\n\
-         #ifdef _WIN64\n\
-         #ifdef __MINGW32__\n\
-         typedef long long value;\n\
-         #else\n\
-         typedef __int64 value;\n\
-         #endif\n\
-         #else\n\
-         typedef long value;\n\
-         #endif\n";
+         #endif";
+         output_without_guarded_primitives poc
+           "\n#include <caml/mlvalues.h>";
+         output_char poc '\n';
          Symtable.output_primitive_table poc;
          output_string poc "\
          #ifdef __cplusplus\n\
diff --git a/bytecomp/dll.ml b/bytecomp/dll.ml
index 0f45d15b75cfa63171802801b3ac29a1184a2ad9..ecdfaefa9c1f8bb86c2daf420752132a114883f2 100644
--- a/bytecomp/dll.ml
+++ b/bytecomp/dll.ml
@@ -196,3 +196,5 @@ let reset () =
   opened_dlls :=[];
   names_of_opened_dlls := [];
   linking_in_core := false
+
+let search_path () = !search_path
diff --git a/bytecomp/dll.mli b/bytecomp/dll.mli
index 5d80e1d4bef5af148c0c4ecbf5f9119d9bbc3110..c93a5b3452887b98c5d9fd12082f1cf8d433a5e1 100644
--- a/bytecomp/dll.mli
+++ b/bytecomp/dll.mli
@@ -55,6 +55,9 @@ val add_path: string list -> unit
 (* Remove the given directories from the search path for DLLs *)
 val remove_path: string list -> unit
 
+(* Return the current search path for DLLs *)
+val search_path: unit -> string list
+
 (* Initialization for separate compilation.
    Initialize the DLL search path to the directories given in the
    environment variable CAML_LD_LIBRARY_PATH, plus contents of ld.conf file
diff --git a/driver/compmisc.ml b/driver/compmisc.ml
index bc2d151b126a770f5ca11354e0e951004c30030d..42479045bfab2b89c93385e9fb34da9bf666b1a6 100644
--- a/driver/compmisc.ml
+++ b/driver/compmisc.ml
@@ -19,21 +19,28 @@
    then the standard library directory (unless the -nostdlib option is given).
  *)
 
-let init_path ?(dir="") () =
+let reinit_path ?(standard_library=Config.standard_library) ?(dir="") () =
   let dirs =
     if !Clflags.use_threads then "+threads" :: !Clflags.include_dirs
     else
       !Clflags.include_dirs
   in
   let dirs =
-    !Compenv.last_include_dirs @ dirs @ Config.flexdll_dirs @
+    !Compenv.last_include_dirs @ dirs @
+    (* Config.flexdll_dirs is either [] or ["+flexdll"]: don't include a
+       reference to the Standard Library when -nostdlib was specified. *)
+    (if !Clflags.no_std_include then [] else Config.flexdll_dirs) @
     !Compenv.first_include_dirs
   in
   let exp_dirs =
-    List.map (Misc.expand_directory Config.standard_library) dirs in
-  Load_path.init (dir :: List.rev_append exp_dirs (Clflags.std_include_dir ()));
+    List.map (Misc.expand_directory standard_library) dirs in
+  let std_include =
+    if !Clflags.no_std_include then [] else [standard_library] in
+  Load_path.init (dir :: List.rev_append exp_dirs std_include);
   Env.reset_cache ()
 
+let init_path = reinit_path ?standard_library:None
+
 (* Return the initial environment in which compilation proceeds. *)
 
 (* Note: do not do init_path() in initial_env, this breaks
diff --git a/driver/compmisc.mli b/driver/compmisc.mli
index bb4c292b4f1e690811b8ed6cd35f98ff429dbe48..67a283b7605af368002286056c7a538c325c8c01 100644
--- a/driver/compmisc.mli
+++ b/driver/compmisc.mli
@@ -14,6 +14,7 @@
 (**************************************************************************)
 
 val init_path : ?dir:string -> unit -> unit
+val reinit_path : ?standard_library:string -> ?dir:string -> unit -> unit
 val initial_env : unit -> Env.t
 
 (* Support for flags that can also be set from an environment variable *)
diff --git a/runtime/Makefile b/runtime/Makefile
index b62a5791a93a22b80f888b34dcdf989eada790c3..9b500d7c8c60a86af71757dbfa4cfde291f5ed8b 100644
--- a/runtime/Makefile
+++ b/runtime/Makefile
@@ -262,9 +262,12 @@ sak.$(O): sak.c caml/misc.h caml/config.h
 C_LITERAL = $(shell ./sak$(EXE) encode-C-literal '$(1)')
 
 build_config.h: $(ROOTDIR)/Makefile.config sak$(EXE)
-	echo '/* This file is generated from $(ROOTDIR)/Makefile.config */' > $@
-	echo '#define OCAML_STDLIB_DIR $(call C_LITERAL,$(LIBDIR))' >> $@
-	echo '#define HOST "$(HOST)"' >> $@
+	{ \
+	  echo '/* This file is generated from $(ROOTDIR)/Makefile.config */'; \
+	  printf '#define OCAML_STDLIB_DIR %s\n' \
+	         '$(call C_LITERAL,$(TARGET_LIBDIR))'; \
+	  echo '#define HOST "$(HOST)"'; \
+	} > $@
 
 # Libraries and programs
 
diff --git a/runtime/caml/exec.h b/runtime/caml/exec.h
index 3955de355ebed81a964f650e594f1f73c2245fd1..66f10b6ffd12235c6014c8b1929810313491adc7 100644
--- a/runtime/caml/exec.h
+++ b/runtime/caml/exec.h
@@ -20,6 +20,8 @@
 
 #ifdef CAML_INTERNALS
 
+#include <stdint.h>
+
 /* Executable bytecode files are composed of a number of sections,
    identified by 4-character names.  A table of contents at the
    end of the file lists the section names along with their sizes,
diff --git a/runtime/caml/memory.h b/runtime/caml/memory.h
index d9e58bc2d097b5dcdc43203f982f6e0db129d8fb..58552eb7cd2406eb83bb1d8b9fdb2afaf579cb4f 100644
--- a/runtime/caml/memory.h
+++ b/runtime/caml/memory.h
@@ -168,6 +168,15 @@ CAMLextern wchar_t* caml_stat_wcsdup(const wchar_t *s);
 */
 CAMLextern caml_stat_string caml_stat_strdup_noexc(const char *s);
 
+/* [caml_stat_memdup(s, size, &out_size)] returns a copy of the first [size]
+   bytes of [s] (which may include NUL characters). If [out_size] is not [NULL],
+   then [size] is stored in [*out_size]. This function is the "dummy" Unix
+   implementation of the Windows-only functions
+   caml_stat_char_array_{to,from}_utf16.
+*/
+CAMLextern caml_stat_string caml_stat_memdup(const char *s, asize_t size,
+                                             asize_t *out_size);
+
 /* [caml_stat_strconcat(nargs, strings)] concatenates NULL-terminated [strings]
    (an array of [char*] of size [nargs]) into a new string, dropping all NULLs,
    except for the very last one. It throws an OCaml exception in case the
diff --git a/runtime/caml/misc.h b/runtime/caml/misc.h
index c605f8711e81fc7411afd529becbaa2c1e9599fb..cfc0337fa5285305a22f63fa25f72b7e08e09e12 100644
--- a/runtime/caml/misc.h
+++ b/runtime/caml/misc.h
@@ -340,6 +340,8 @@ extern double caml_log1p(double);
 
 #define caml_stat_strdup_to_os caml_stat_strdup_to_utf16
 #define caml_stat_strdup_of_os caml_stat_strdup_of_utf16
+#define caml_stat_char_array_to_os caml_stat_char_array_to_utf16
+#define caml_stat_char_array_of_os caml_stat_char_array_of_utf16
 #define caml_copy_string_of_os caml_copy_string_of_utf16
 
 #else /* _WIN32 */
@@ -380,6 +382,8 @@ extern double caml_log1p(double);
 
 #define caml_stat_strdup_to_os caml_stat_strdup
 #define caml_stat_strdup_of_os caml_stat_strdup
+#define caml_stat_char_array_to_os caml_stat_memdup
+#define caml_stat_char_array_of_os caml_stat_memdup
 #define caml_copy_string_of_os caml_copy_string
 
 #endif /* _WIN32 */
diff --git a/runtime/caml/osdeps.h b/runtime/caml/osdeps.h
index bc8cd3b9e682b93b295b700d3ee312243734c839..f43bf6f73dc3fb5ac694e17776c5ded6f0bd629a 100644
--- a/runtime/caml/osdeps.h
+++ b/runtime/caml/osdeps.h
@@ -142,25 +142,68 @@ CAMLextern clock_t caml_win32_clock(void);
 
 #ifdef _WIN32
 
-/* [caml_stat_strdup_to_utf16(s)] returns a NULL-terminated copy of [s],
+/* [caml_stat_strdup_to_utf16(s)] returns a NUL-terminated copy of [s],
    re-encoded in UTF-16.  The encoding of [s] is assumed to be UTF-8 if
    [caml_windows_unicode_runtime_enabled] is non-zero **and** [s] is valid
    UTF-8, or the current Windows code page otherwise.
 
-   The returned string is allocated with [caml_stat_alloc], so it should be free
-   using [caml_stat_free].
+   The returned string is allocated with [caml_stat_alloc], so it should be
+   freed using [caml_stat_free].
+
+   If allocation fails, this raises Out_of_memory. This function may also raise
+   [Sys_error] if [s] is not valid UTF-8.
 */
 CAMLextern wchar_t* caml_stat_strdup_to_utf16(const char *s);
 
-/* [caml_stat_strdup_of_utf16(s)] returns a NULL-terminated copy of [s],
+/* [caml_stat_strdup_of_utf16(s)] returns a NUL-terminated copy of [s],
    re-encoded in UTF-8 if [caml_windows_unicode_runtime_enabled] is non-zero or
    the current Windows code page otherwise.
 
-   The returned string is allocated with [caml_stat_alloc], so it should be free
-   using [caml_stat_free].
+   The returned string is allocated with [caml_stat_alloc], so it should be
+   freed using [caml_stat_free].
+
+   If allocation fails, this raises Out_of_memory. This function may also raise
+   [Sys_error] if [s] is not valid UTF-16.
 */
 CAMLextern char* caml_stat_strdup_of_utf16(const wchar_t *s);
 
+/* [caml_stat_char_array_to_utf16(s, size, &out_size)] returns a copy of the
+   first [size] bytes of [s] re-encoded in UTF-16. [s] does not have to be NUL-
+   terminated and may contain embedded NUL characters. The encoding of [s] is
+   assumed to be UTF-8 if [caml_windows_unicode_runtime_enabled] is non-zero
+   **and** [s] is valid UTF-8, or the current Windows code page otherwise. If
+   [out_size] is not [NULL], then the number of UTF-16 code units in the result
+   is recorded in [*out_size].
+
+   [size] must be greater than zero.
+
+   The returned buffer is allocated with [caml_stat_alloc], so it should be
+   freed using [caml_stat_free].
+
+   If allocation fails, this raises Out_of_memory. This function may also raise
+   [Sys_error] if [s] is not valid UTF-8.
+*/
+CAMLextern wchar_t *caml_stat_char_array_to_utf16(const char *s, size_t size,
+                                                  size_t *out_size);
+
+/* [caml_stat_char_array_of_utf16(s, size, &out_size)] returns a copy of the
+   first [size] UTF-16 code units of [s] re-encoded in UTF-8 if
+   [caml_windows_unicode_runtime_enabled] is non-zero or the current Windows
+   code page otherwise. [s] does not have to be NUL-terminated and may contain
+   embedded NUL characters. If [out_size] is not [NULL], then the size of the
+   result in bytes recorded in [*out_size].
+
+   [size] must be greater than zero.
+
+   The returned buffer is allocated with [caml_stat_alloc], so it should be
+   freed using [caml_stat_free].
+
+   If allocation fails, this raises Out_of_memory. This function may also raise
+   [Sys_error] if [s] is not valid UTF-16.
+*/
+CAMLextern char *caml_stat_char_array_of_utf16(const wchar_t *s, size_t size,
+                                               size_t *out_size);
+
 /* [caml_copy_string_of_utf16(s)] returns an OCaml string containing a copy of
    [s] re-encoded in UTF-8 if [caml_windows_unicode_runtime_enabled] is non-zero
    or in the current code page otherwise.
diff --git a/runtime/memory.c b/runtime/memory.c
index 66d1c50ccb2d6ae2618e40dacc9de54e9ed9ea68..e85883de8035a7c99087199dc20280285dcc92a7 100644
--- a/runtime/memory.c
+++ b/runtime/memory.c
@@ -904,6 +904,18 @@ CAMLexport caml_stat_string caml_stat_strdup(const char *s)
   return result;
 }
 
+CAMLexport caml_stat_string caml_stat_memdup(const char *s, asize_t size,
+                                             asize_t *out_size)
+{
+  caml_stat_block result;
+  CAMLassert(size > 0);
+  result = caml_stat_alloc(size);
+  memcpy(result, s, size);
+  if (out_size != NULL)
+    *out_size = size;
+  return result;
+}
+
 #ifdef _WIN32
 
 CAMLexport wchar_t * caml_stat_wcsdup(const wchar_t *s)
diff --git a/runtime/startup_byt.c b/runtime/startup_byt.c
index de7549748ac10752d03e00ba6f5e1ce55389495a..15b7aded01e7f9849deffbcf3c003bc3236b8087 100644
--- a/runtime/startup_byt.c
+++ b/runtime/startup_byt.c
@@ -427,6 +427,8 @@ static void do_print_config(void)
 
   /* Parse ld.conf and print the effective search path */
   puts("shared_libs_path:");
+  caml_decompose_path(&caml_shared_libs_path,
+                      caml_secure_getenv(T("CAML_LD_LIBRARY_PATH")));
   caml_parse_ld_conf();
   for (i = 0; i < caml_shared_libs_path.size; i++) {
     dir = caml_shared_libs_path.contents[i];
diff --git a/runtime/win32.c b/runtime/win32.c
index 77e5f39fb057fb8c4e11d1d4d56ebfad5652da4b..a11ddb5d6f81c7d58f57061c97476257e6822db4 100644
--- a/runtime/win32.c
+++ b/runtime/win32.c
@@ -930,30 +930,47 @@ CAMLexport value caml_copy_string_of_utf16(const wchar_t *s)
   return v;
 }
 
-CAMLexport wchar_t* caml_stat_strdup_to_utf16(const char *s)
+CAMLexport wchar_t *caml_stat_char_array_to_utf16(const char *s, size_t size,
+                                                  size_t *out_size)
 {
   wchar_t * ws;
   int retcode;
 
-  retcode = win_multi_byte_to_wide_char(s, -1, NULL, 0);
-  ws = caml_stat_alloc_noexc(retcode * sizeof(*ws));
-  win_multi_byte_to_wide_char(s, -1, ws, retcode);
+  retcode = win_multi_byte_to_wide_char(s, size, NULL, 0);
+  ws = caml_stat_alloc(retcode * sizeof(wchar_t));
+  win_multi_byte_to_wide_char(s, size, ws, retcode);
+  if (out_size != NULL)
+    *out_size = retcode;
 
   return ws;
 }
 
-CAMLexport caml_stat_string caml_stat_strdup_of_utf16(const wchar_t *s)
+CAMLexport wchar_t *caml_stat_strdup_to_utf16(const char *s)
+{
+  return caml_stat_char_array_to_utf16(s, -1, NULL);
+}
+
+CAMLexport caml_stat_string caml_stat_char_array_of_utf16(const wchar_t *s,
+                                                          size_t size,
+                                                          size_t *out_size)
 {
   caml_stat_string out;
   int retcode;
 
-  retcode = win_wide_char_to_multi_byte(s, -1, NULL, 0);
+  retcode = win_wide_char_to_multi_byte(s, size, NULL, 0);
   out = caml_stat_alloc(retcode);
-  win_wide_char_to_multi_byte(s, -1, out, retcode);
+  win_wide_char_to_multi_byte(s, size, out, retcode);
+  if (out_size != NULL)
+    *out_size = retcode;
 
   return out;
 }
 
+CAMLexport caml_stat_string caml_stat_strdup_of_utf16(const wchar_t *s)
+{
+  return caml_stat_char_array_of_utf16(s, -1, NULL);
+}
+
 void caml_probe_win32_version(void)
 {
   /* Determine the version of Windows we're running, and cache it */
diff --git a/stdlib/Makefile b/stdlib/Makefile
index 0f2594e9c72e9ef24d451403f1dfb2bc634470bc..7810e9edf082b9a33b25432b0891cd439fe49b1e 100644
--- a/stdlib/Makefile
+++ b/stdlib/Makefile
@@ -17,8 +17,6 @@ ROOTDIR = ..
 
 include $(ROOTDIR)/Makefile.common
 
-TARGET_BINDIR ?= $(BINDIR)
-
 # There are three ways the Standard Library is compiled in bytecode:
 # 1. During coldstart
 #      - using ../boot/ocamlc which runs on ../boot/ocamlrun
@@ -57,14 +55,14 @@ NOSTDLIB= camlinternalFormatBasics.cmo camlinternalAtomic.cmo stdlib.cmo
 OTHERS=$(filter-out $(NOSTDLIB),$(OBJS))
 
 .PHONY: all
-all: stdlib.cma std_exit.cmo camlheader target_camlheader camlheader_ur
+all: stdlib.cma std_exit.cmo $(HEADER_NAME) target_$(HEADER_NAME) camlheader_ur
 
 ifeq "$(RUNTIMED)" "true"
-all: camlheaderd target_camlheaderd
+all: target_camlheaderd
 endif
 
 ifeq "$(INSTRUMENTED_RUNTIME)" "true"
-all: camlheaderi target_camlheaderi
+all: target_camlheaderi
 endif
 
 .PHONY: allopt opt.opt # allopt and opt.opt are synonyms
@@ -91,7 +89,7 @@ ifeq "$(INSTALL_SOURCE_ARTIFACTS)" "true"
 	  *.cmt *.cmti *.mli *.ml \
 	  "$(INSTALL_LIBDIR)"
 endif
-	$(INSTALL_DATA) target_camlheader "$(INSTALL_LIBDIR)/camlheader"
+	$(INSTALL_DATA) target_$(HEADER_NAME) "$(INSTALL_LIBDIR)/$(HEADER_NAME)"
 
 ifeq "$(RUNTIMED)" "true"
 install::
@@ -122,92 +120,71 @@ installopt-default::
 	  stdlib.cmxa stdlib.$(A) std_exit.$(O) *.cmx \
 	  "$(INSTALL_LIBDIR)"
 
-ifeq "$(UNIX_OR_WIN32)" "unix"
-HEADERPROGRAM = header
-HEADER_PATH = $(BINDIR)/
-HEADER_TARGET_PATH = $(TARGET_BINDIR)/
-else # Windows
-HEADERPROGRAM = headernt
-HEADER_PATH =
-HEADER_TARGET_PATH =
-endif
-
-TARGETHEADERPROGRAM = target_$(HEADERPROGRAM)
-
-# The shebang test in configure.ac will need updating if any runtime is
-# introduced with a suffix more than one character long (camlheader_ur doesn't
-# matter).
-CAMLHEADERS =\
-  camlheader target_camlheader camlheader_ur \
-  camlheaderd target_camlheaderd \
-  camlheaderi target_camlheaderi
-
-# The % in pattern rules must always match something, hence the slightly strange
-# patterns and $(subst ...) since `camlheader%:` wouldn't match `camlheader`
 ifeq "$(SHEBANGSCRIPTS)" "true"
-camlhead%: $(ROOTDIR)/Makefile.config Makefile
-ifeq "$(LONG_SHEBANG)" "true"
-	echo '#!/bin/sh' > $@
-	echo 'exec "$(BINDIR)/ocamlrun$(subst er,,$*)" "$$0" "$$@"' >> $@
-else
-	echo '#!$(BINDIR)/ocamlrun$(subst er,,$*)' > $@
-endif
-
-# TODO This does not take long shebangs into account (since TARGET_BINDIR is not
-#      yet processed by configure)
-target_%: $(ROOTDIR)/Makefile.config Makefile
-	echo '#!$(TARGET_BINDIR)/ocamlrun$(subst camlheader,,$*)' > $@
 
 camlheader_ur: Makefile
 	echo '#!' | tr -d '\012' > $@
 
+target_camlheader%: Makefile
+	echo '#!/usr/bin/false' > $@
+
 else # Hashbang scripts not supported
 
-$(CAMLHEADERS): $(HEADERPROGRAM).c $(ROOTDIR)/Makefile.config Makefile
+camlheader_ur: tmpheader.exe
+	rm -f $@.exe
+	cp $< $@
 
-# $@.exe is deleted to ensure no Cygwin .exe mangling takes place
-camlhead%: tmphead%.exe
+target_camlheader%: tmpheader.exe
 	rm -f $@.exe
-	mv $< $@
+	cp $< $@
 
-# Again, pattern weirdness here means that the dot is always present so that
-# tmpheader.exe matches.
-tmpheader%exe: $(HEADERPROGRAM)%$(O)
-	$(call MKEXE_USING_COMPILER,$@,$^ $(EXTRALIBS))
-# FIXME This is wrong - mingw could invoke strip; MSVC equivalent?
-ifneq "$(UNIX_OR_WIN32)" "win32"
-	strip $@
 endif
 
-$(HEADERPROGRAM)%$(O): \
-  OC_CPPFLAGS += -DRUNTIME_NAME='"$(HEADER_PATH)ocamlrun$(subst .,,$*)"'
-
-$(HEADERPROGRAM)%$(O): $(HEADERPROGRAM).c
-	$(CC) -c $(OC_CFLAGS) $(CFLAGS) $(OC_CPPFLAGS) $(CPPFLAGS) \
-	  $(OUTPUTOBJ)$@ $^
-
-camlheader_ur: camlheader
-	cp camlheader $@
-
-ifeq "$(UNIX_OR_WIN32)" "unix"
-tmptargetcamlheader%exe: $(TARGETHEADERPROGRAM)%$(O)
-	$(call MKEXE_USING_COMPILER,$@,$^ $(EXTRALIBS))
-	strip $@
-
-$(TARGETHEADERPROGRAM)%$(O): $(HEADERPROGRAM).c
-	$(CC) -c $(OC_CFLAGS) $(CFLAGS) $(OC_CPPFLAGS) $(CPPFLAGS) \
-	      -DRUNTIME_NAME='"$(HEADER_TARGET_PATH)ocamlrun$(subst .,,$*)"' \
-	      $(OUTPUTOBJ)$@ $^
-
-target_%: tmptarget%.exe
-	rm -f $@.exe
-	mv $< $@
+target_$(HEADER_NAME): target_runtime.info tmpheader.exe
+	@cat $^ > $@
+
+$(HEADER_NAME): runtime.info tmpheader.exe
+	@cat $^ > $@
+
+# The mingw-w64 and MSVC versions of tmpheader.exe are linked with special flags
+# to reduce their size (considerably). In particular, the entry point is
+# overridden, which prevents the linking of crt0.
+ifeq "$(TOOLCHAIN)" "mingw"
+# mingw-w64: optimise header.o for space and remove all unused sections during
+# linking.
+header.o: OC_CFLAGS += -Os
+ifeq "$(SYSTEM)" "mingw"
+  ENTRYPOINT = _wmainCRTStartup
 else
-target_%: %
-	cp $< $@
+  ENTRYPOINT = wmainCRTStartup
+endif
+# Certainly for GCC, -nostdlib automatically adds -nostartfiles (clang's
+# documentation is not explicit on this point, but it would be unusual for clang
+# and GCC to differ). --gc-sections removes unreferenced code and data sections.
+tmpheader.exe: OC_LDFLAGS += \
+  -nostdlib -nostartfiles -Wl,--gc-sections -Wl,-e,$(ENTRYPOINT)
+HEADERLIBS = -lkernel32
+else ifeq "$(TOOLCHAIN)" "msvc"
+# MSVC: Optimise header.obj for space. cl doesn't have an equivalent of
+# -nostdlib, however it also doesn't have default libraries in the same way as
+# GCC - the equivalent of crt0.o is linked because it provides the entry point
+# symbol, rather than by being explicitly added to the linking command line.
+# /GS- disables compilation of runtime security checks and /NOCOFFGRPINFO is an
+# undocumented linker flag which removes the debug directory from the PE+ image.
+header.obj: OC_CFLAGS += /GS- /Os
+tmpheader.exe: OC_LDFLAGS += /NOCOFFGRPINFO /subsystem:console
+HEADERLIBS = kernel32.lib
+else
+HEADERLIBS =
 endif
 
-endif # ifeq "$(SHEBANGSCRIPTS)" "true"
+.INTERMEDIATE: tmpheader.exe
+tmpheader.exe: header.$(O)
+	$(call MKEXE_USING_COMPILER,$@,$^ $(HEADERLIBS))
+# FIXME This is wrong - mingw could invoke strip; MSVC equivalent?
+ifneq "$(UNIX_OR_WIN32)" "win32"
+	strip $@
+endif
 
 stdlib.cma: $(OBJS)
 	$(CAMLC) -a -o $@ $^
@@ -217,11 +194,11 @@ stdlib.cmxa: $(OBJS:.cmo=.cmx)
 
 .PHONY: distclean
 distclean: clean
-	rm -f sys.ml
+	rm -f sys.ml runtime.info target_runtime.info
 
 .PHONY: clean
 clean::
-	rm -f $(CAMLHEADERS)
+	rm -f $(HEADER_NAME) target_$(HEADER_NAME)* camlheader_ur
 
 export AWK
 
diff --git a/stdlib/header.c b/stdlib/header.c
index 426b064e2fdb439b3cfd43af5770c089def2e235..30c7be7048c6b8233b72c441e8b84209b8b5f2c7 100644
--- a/stdlib/header.c
+++ b/stdlib/header.c
@@ -13,36 +13,102 @@
 /*                                                                        */
 /**************************************************************************/
 
-#define CAML_INTERNALS
+/* The launcher for bytecode executables (if #! is not available) */
+
+/* C11's _Noreturn is deprecated in C23 in favour of attributes */
+#if defined(__STDC_VERSION__) && __STDC_VERSION__ >= 202311L
+  #define NORETURN [[noreturn]]
+#elif defined(__STDC_VERSION__) && __STDC_VERSION__ >= 201112L
+  #define NORETURN _Noreturn
+#elif defined(_MSC_VER)
+  #define NORETURN __declspec(noreturn)
+#else
+  #define NORETURN
+#endif
+
+#ifdef _WIN32
+
+#define STRICT
+#define WIN32_LEAN_AND_MEAN
+#include <windows.h>
+
+#if WINDOWS_UNICODE
+#define CP CP_UTF8
+#else
+#define CP CP_ACP
+#endif
+
+/* mingw-w64 has a limits.h which defines PATH_MAX as an alias for MAX_PATH */
+#if !defined(PATH_MAX)
+#define PATH_MAX MAX_PATH
+#endif
+
+#define SEEK_END FILE_END
+
+#define lseek(h, offset, origin) SetFilePointer((h), (offset), NULL, (origin))
+
+typedef HANDLE file_descriptor;
+
+static int read(HANDLE h, LPVOID buffer, DWORD buffer_size)
+{
+  DWORD nread = 0;
+  ReadFile(h, buffer, buffer_size, &nread, NULL);
+  return nread;
+}
+
+static BOOL WINAPI ctrl_handler(DWORD event)
+{
+  if (event == CTRL_C_EVENT || event == CTRL_BREAK_EVENT)
+    return TRUE;                /* pretend we've handled them */
+  else
+    return FALSE;
+}
 
-/* The launcher for bytecode executables (if #! is not working) */
+static void write_error(const wchar_t *wstr, HANDLE hOut)
+{
+  DWORD consoleMode, numwritten, len;
+  char str[MAX_PATH];
+
+  if (GetConsoleMode(hOut, &consoleMode) != 0) {
+    /* The output stream is a Console */
+    WriteConsole(hOut, wstr, lstrlen(wstr), &numwritten, NULL);
+  } else { /* The output stream is redirected */
+    len =
+      WideCharToMultiByte(CP, 0, wstr, lstrlen(wstr), str, sizeof(str),
+                          NULL, NULL);
+    WriteFile(hOut, str, len, &numwritten, NULL);
+  }
+}
+
+NORETURN static void exit_with_error(const wchar_t *wstr1,
+                                     const wchar_t *wstr2,
+                                     const wchar_t *wstr3)
+{
+  HANDLE hOut = GetStdHandle(STD_ERROR_HANDLE);
+  if (wstr1) write_error(wstr1, hOut);
+  if (wstr2) write_error(wstr2, hOut);
+  if (wstr3) write_error(wstr3, hOut);
+  write_error(L"\r\n", hOut);
+  ExitProcess(2);
+}
+
+#else
 
 #include <stdio.h>
 #include <stdlib.h>
 #include <string.h>
-#include "caml/s.h"
-#ifdef HAS_UNISTD
 #include <unistd.h>
-#endif
 #include <fcntl.h>
+#include <limits.h>
 #include <sys/types.h>
 #include <sys/stat.h>
-#include "caml/mlvalues.h"
-#include "caml/exec.h"
-
-char * default_runtime_path = RUNTIME_NAME;
-
-#ifndef MAXPATHLEN
-#define MAXPATHLEN 1024
-#endif
 
-#ifndef S_ISREG
-#define S_ISREG(mode) (((mode) & S_IFMT) == S_IFREG)
+/* O_BINARY is defined in Gnulib, but is not POSIX */
+#ifndef O_BINARY
+#define O_BINARY 0
 #endif
 
-#ifndef SEEK_END
-#define SEEK_END 2
-#endif
+typedef int file_descriptor;
 
 #ifndef __CYGWIN__
 
@@ -50,24 +116,23 @@ char * default_runtime_path = RUNTIME_NAME;
 
 static char * searchpath(char * name)
 {
-  static char fullname[MAXPATHLEN + 1];
+  static char fullname[PATH_MAX + 1];
   char * path;
-  char * p;
-  char * q;
   struct stat st;
 
-  for (p = name; *p != 0; p++) {
+  for (char *p = name; *p != 0; p++) {
     if (*p == '/') return name;
   }
   path = getenv("PATH");
   if (path == NULL) return name;
   while(1) {
+    char * p;
     for (p = fullname; *path != 0 && *path != ':'; p++, path++)
-      if (p < fullname + MAXPATHLEN) *p = *path;
-    if (p != fullname && p < fullname + MAXPATHLEN)
+      if (p < fullname + PATH_MAX) *p = *path;
+    if (p != fullname && p < fullname + PATH_MAX)
       *p++ = '/';
-    for (q = name; *q != 0; p++, q++)
-      if (p < fullname + MAXPATHLEN) *p = *q;
+    for (char *q = name; *q != 0; p++, q++)
+      if (p < fullname + PATH_MAX) *p = *q;
     *p = 0;
     if (stat(fullname, &st) == 0 && S_ISREG(st.st_mode)) break;
     if (*path == 0) return name;
@@ -92,14 +157,14 @@ static int file_ok(char * name)
 
 static char * searchpath(char * name)
 {
-  char * path, * fullname, * p;
+  char * path, * fullname;
 
   path = getenv("PATH");
   fullname = malloc(strlen(name) + (path == NULL ? 0 : strlen(path)) + 6);
   /* 6 = "/" plus ".exe" plus final "\0" */
   if (fullname == NULL) return name;
   /* Check for absolute path name */
-  for (p = name; *p != 0; p++) {
+  for (char *p = name; *p != 0; p++) {
     if (*p == '/' || *p == '\\') {
       if (file_ok(name)) return name;
       strcpy(fullname, name);
@@ -111,6 +176,7 @@ static char * searchpath(char * name)
   /* Search in path */
   if (path == NULL) return name;
   while(1) {
+    char * p;
     for (p = fullname; *path != 0 && *path != ':'; p++, path++) *p = *path;
     if (p != fullname) *p++ = '/';
     strcpy(p, name);
@@ -125,26 +191,43 @@ static char * searchpath(char * name)
 
 #endif
 
-static unsigned long read_size(char * ptr)
+NORETURN static void exit_with_error(const char *str1,
+                                     const char *str2,
+                                     const char *str3)
 {
-  unsigned char * p = (unsigned char *) ptr;
-  return ((unsigned long) p[0] << 24) + ((unsigned long) p[1] << 16) +
-         ((unsigned long) p[2] << 8) + p[3];
+  if (str1) fputs(str1, stderr);
+  if (str2) fputs(str2, stderr);
+  if (str3) fputs(str3, stderr);
+  fputs("\n", stderr);
+  exit(2);
 }
 
-static char * read_runtime_path(int fd)
+#endif /* defined(_WIN32) */
+
+#define CAML_INTERNALS
+#include "caml/exec.h"
+
+static uint32_t read_size(const char *ptr)
+{
+  const unsigned char *p = (const unsigned char *)ptr;
+  return ((uint32_t) p[0] << 24) | ((uint32_t) p[1] << 16) |
+         ((uint32_t) p[2] << 8) | p[3];
+}
+
+static char * read_runtime_path(file_descriptor fd)
 {
   char buffer[TRAILER_SIZE];
-  static char runtime_path[MAXPATHLEN];
-  int num_sections, i;
+  static char runtime_path[PATH_MAX];
+  int num_sections;
   uint32_t path_size;
   long ofs;
+  int i;
 
-  lseek(fd, (long) -TRAILER_SIZE, SEEK_END);
+  if (lseek(fd, -TRAILER_SIZE, SEEK_END) == -1) return NULL;
   if (read(fd, buffer, TRAILER_SIZE) < TRAILER_SIZE) return NULL;
   num_sections = read_size(buffer);
   ofs = TRAILER_SIZE + num_sections * 8;
-  lseek(fd, -ofs, SEEK_END);
+  if (lseek(fd, -ofs, SEEK_END) == -1) return NULL;
   path_size = 0;
   for (i = 0; i < num_sections; i++) {
     if (read(fd, buffer, 8) < 8) return NULL;
@@ -155,39 +238,81 @@ static char * read_runtime_path(int fd)
     } else if (path_size > 0)
       ofs += read_size(buffer + 4);
   }
-  if (path_size == 0) return default_runtime_path;
-  if (path_size >= MAXPATHLEN) return NULL;
-  lseek(fd, -ofs, SEEK_END);
+  if (path_size == 0) return NULL;
+  if (path_size >= PATH_MAX) return NULL;
+  if (lseek(fd, -ofs, SEEK_END) == -1) return NULL;
   if (read(fd, runtime_path, path_size) != path_size) return NULL;
-  runtime_path[path_size - 1] = 0;
   return runtime_path;
 }
 
-static void errwrite(char * msg)
+#ifdef _WIN32
+
+NORETURN void __cdecl wmainCRTStartup(void)
 {
-  write(2, msg, strlen(msg));
+  wchar_t truename[MAX_PATH];
+  char *runtime_path;
+  wchar_t wruntime_path[MAX_PATH];
+  HANDLE h;
+  STARTUPINFO stinfo;
+  PROCESS_INFORMATION procinfo;
+  DWORD retcode;
+
+  if (GetModuleFileName(NULL, truename, sizeof(truename)/sizeof(wchar_t)) == 0)
+    exit_with_error(L"Out of memory", NULL, NULL);
+
+  h = CreateFile(truename, GENERIC_READ, FILE_SHARE_READ | FILE_SHARE_WRITE,
+                 NULL, OPEN_EXISTING, 0, NULL);
+  if (h == INVALID_HANDLE_VALUE ||
+      (runtime_path = read_runtime_path(h)) == NULL ||
+      !MultiByteToWideChar(CP, 0, runtime_path, -1, wruntime_path,
+                           sizeof(wruntime_path)/sizeof(wchar_t)))
+    exit_with_error(NULL, truename,
+                    L" not found or is not a bytecode executable file");
+  CloseHandle(h);
+  if (SearchPath(NULL, wruntime_path, L".exe", sizeof(truename)/sizeof(wchar_t),
+                 truename, NULL)) {
+    /* Need to ignore ctrl-C and ctrl-break, otherwise we'll die and take
+       the underlying OCaml program with us! */
+    SetConsoleCtrlHandler(ctrl_handler, TRUE);
+
+    stinfo.cb = sizeof(stinfo);
+    stinfo.lpReserved = NULL;
+    stinfo.lpDesktop = NULL;
+    stinfo.lpTitle = NULL;
+    stinfo.dwFlags = 0;
+    stinfo.cbReserved2 = 0;
+    stinfo.lpReserved2 = NULL;
+    if (CreateProcess(truename, GetCommandLine(), NULL, NULL, TRUE, 0,
+                      NULL, NULL, &stinfo, &procinfo)) {
+      CloseHandle(procinfo.hThread);
+      WaitForSingleObject(procinfo.hProcess, INFINITE);
+      GetExitCodeProcess(procinfo.hProcess, &retcode);
+      CloseHandle(procinfo.hProcess);
+      ExitProcess(retcode);
+    }
+  }
+
+  exit_with_error(L"Cannot exec ", wruntime_path, NULL);
 }
 
-#ifndef O_BINARY
-#define O_BINARY 0
-#endif
+#else
 
-int main(int argc, char ** argv)
+int main(int argc, char *argv[])
 {
-  char * truename, * runtime_path;
+  char *truename, *runtime_path;
   int fd;
 
   truename = searchpath(argv[0]);
   fd = open(truename, O_RDONLY | O_BINARY);
-  if (fd == -1 || (runtime_path = read_runtime_path(fd)) == NULL) {
-    errwrite(truename);
-    errwrite(" not found or is not a bytecode executable file\n");
-    return 2;
-  }
+  if (fd == -1 || (runtime_path = read_runtime_path(fd)) == NULL)
+    exit_with_error(NULL, truename,
+                    " not found or is not a bytecode executable file");
+  close(fd);
+
   argv[0] = truename;
-  execv(runtime_path, argv);
-  errwrite("Cannot exec ");
-  errwrite(runtime_path);
-  errwrite("\n");
-  return 2;
+  execvp(runtime_path, argv);
+
+  exit_with_error("Cannot exec ", runtime_path, NULL);
 }
+
+#endif /* defined(_WIN32) */
diff --git a/stdlib/headernt.c b/stdlib/headernt.c
deleted file mode 100644
index 43109c865ea25b1304269873e6cf3f931e279e62..0000000000000000000000000000000000000000
--- a/stdlib/headernt.c
+++ /dev/null
@@ -1,196 +0,0 @@
-/**************************************************************************/
-/*                                                                        */
-/*                                 OCaml                                  */
-/*                                                                        */
-/*             Xavier Leroy, projet Cristal, INRIA Rocquencourt           */
-/*                                                                        */
-/*   Copyright 1998 Institut National de Recherche en Informatique et     */
-/*     en Automatique.                                                    */
-/*                                                                        */
-/*   All rights reserved.  This file is distributed under the terms of    */
-/*   the GNU Lesser General Public License version 2.1, with the          */
-/*   special exception on linking described in the file LICENSE.          */
-/*                                                                        */
-/**************************************************************************/
-
-#define CAML_INTERNALS
-
-#define STRICT
-#define WIN32_LEAN_AND_MEAN
-
-#include <windows.h>
-#include "caml/mlvalues.h"
-#include "caml/exec.h"
-
-#ifndef __MINGW32__
-#pragma comment(linker , "/subsystem:console")
-#pragma comment(lib , "kernel32")
-#ifdef _UCRT
-#pragma comment(lib , "ucrt.lib")
-#pragma comment(lib , "vcruntime.lib")
-#endif
-#endif
-
-char * default_runtime_name = RUNTIME_NAME;
-
-static
-#if _MSC_VER >= 1200
-__forceinline
-#else
-__inline
-#endif
-unsigned long read_size(const char * const ptr)
-{
-  const unsigned char * const p = (const unsigned char * const) ptr;
-  return ((unsigned long) p[0] << 24) | ((unsigned long) p[1] << 16) |
-         ((unsigned long) p[2] << 8) | p[3];
-}
-
-static __inline char * read_runtime_path(HANDLE h)
-{
-  char buffer[TRAILER_SIZE];
-  static char runtime_path[MAX_PATH];
-  DWORD nread;
-  int num_sections, path_size, i;
-  long ofs;
-
-  if (SetFilePointer(h, -TRAILER_SIZE, NULL, FILE_END) == -1) return NULL;
-  if (! ReadFile(h, buffer, TRAILER_SIZE, &nread, NULL)) return NULL;
-  if (nread != TRAILER_SIZE) return NULL;
-  num_sections = read_size(buffer);
-  ofs = TRAILER_SIZE + num_sections * 8;
-  if (SetFilePointer(h, - ofs, NULL, FILE_END) == -1) return NULL;
-  path_size = 0;
-  for (i = 0; i < num_sections; i++) {
-    if (! ReadFile(h, buffer, 8, &nread, NULL) || nread != 8) return NULL;
-    if (buffer[0] == 'R' && buffer[1] == 'N' &&
-        buffer[2] == 'T' && buffer[3] == 'M') {
-      path_size = read_size(buffer + 4);
-      ofs += path_size;
-    } else if (path_size > 0)
-      ofs += read_size(buffer + 4);
-  }
-  if (path_size == 0) return default_runtime_name;
-  if (path_size >= MAX_PATH) return NULL;
-  if (SetFilePointer(h, -ofs, NULL, FILE_END) == -1) return NULL;
-  if (! ReadFile(h, runtime_path, path_size, &nread, NULL)) return NULL;
-  if (nread != path_size) return NULL;
-  runtime_path[path_size - 1] = 0;
-  return runtime_path;
-}
-
-static BOOL WINAPI ctrl_handler(DWORD event)
-{
-  if (event == CTRL_C_EVENT || event == CTRL_BREAK_EVENT)
-    return TRUE;                /* pretend we've handled them */
-  else
-    return FALSE;
-}
-
-#if WINDOWS_UNICODE
-#define CP CP_UTF8
-#else
-#define CP CP_ACP
-#endif
-
-static void write_console(HANDLE hOut, WCHAR *wstr)
-{
-  DWORD consoleMode, numwritten, len;
-  static char str[MAX_PATH];
-
-  if (GetConsoleMode(hOut, &consoleMode) != 0) {
-    /* The output stream is a Console */
-    WriteConsole(hOut, wstr, wcslen(wstr), &numwritten, NULL);
-  } else { /* The output stream is redirected */
-    len =
-      WideCharToMultiByte(CP, 0, wstr, wcslen(wstr), str, sizeof(str),
-                          NULL, NULL);
-    WriteFile(hOut, str, len, &numwritten, NULL);
-  }
-}
-
-static __inline void __declspec(noreturn) run_runtime(wchar_t * runtime,
-         wchar_t * const cmdline)
-{
-  wchar_t path[MAX_PATH];
-  STARTUPINFO stinfo;
-  PROCESS_INFORMATION procinfo;
-  DWORD retcode;
-  if (SearchPath(NULL, runtime, L".exe", sizeof(path)/sizeof(wchar_t),
-                 path, &runtime) == 0) {
-    HANDLE errh;
-    errh = GetStdHandle(STD_ERROR_HANDLE);
-    write_console(errh, L"Cannot exec ");
-    write_console(errh, runtime);
-    write_console(errh, L"\r\n");
-    ExitProcess(2);
-#if _MSC_VER >= 1200
-    __assume(0); /* Not reached */
-#endif
-  }
-  /* Need to ignore ctrl-C and ctrl-break, otherwise we'll die and take
-     the underlying OCaml program with us! */
-  SetConsoleCtrlHandler(ctrl_handler, TRUE);
-
-  stinfo.cb = sizeof(stinfo);
-  stinfo.lpReserved = NULL;
-  stinfo.lpDesktop = NULL;
-  stinfo.lpTitle = NULL;
-  stinfo.dwFlags = 0;
-  stinfo.cbReserved2 = 0;
-  stinfo.lpReserved2 = NULL;
-  if (!CreateProcess(path, cmdline, NULL, NULL, TRUE, 0, NULL, NULL,
-                     &stinfo, &procinfo)) {
-    HANDLE errh;
-    errh = GetStdHandle(STD_ERROR_HANDLE);
-    write_console(errh, L"Cannot exec ");
-    write_console(errh, runtime);
-    write_console(errh, L"\r\n");
-    ExitProcess(2);
-#if _MSC_VER >= 1200
-    __assume(0); /* Not reached */
-#endif
-  }
-  CloseHandle(procinfo.hThread);
-  WaitForSingleObject(procinfo.hProcess , INFINITE);
-  GetExitCodeProcess(procinfo.hProcess , &retcode);
-  CloseHandle(procinfo.hProcess);
-  ExitProcess(retcode);
-#if _MSC_VER >= 1200
-    __assume(0); /* Not reached */
-#endif
-}
-
-int wmain(void)
-{
-  wchar_t truename[MAX_PATH];
-  wchar_t * cmdline = GetCommandLine();
-  char * runtime_path;
-  wchar_t wruntime_path[MAX_PATH];
-  HANDLE h;
-
-  GetModuleFileName(NULL, truename, sizeof(truename)/sizeof(wchar_t));
-  h = CreateFile(truename, GENERIC_READ, FILE_SHARE_READ | FILE_SHARE_WRITE,
-                 NULL, OPEN_EXISTING, 0, NULL);
-  if (h == INVALID_HANDLE_VALUE ||
-      (runtime_path = read_runtime_path(h)) == NULL) {
-    HANDLE errh;
-    errh = GetStdHandle(STD_ERROR_HANDLE);
-    write_console(errh, truename);
-    write_console(errh, L" not found or is not a bytecode executable file\r\n");
-    ExitProcess(2);
-#if _MSC_VER >= 1200
-    __assume(0); /* Not reached */
-#endif
-  }
-  CloseHandle(h);
-  MultiByteToWideChar(CP, 0, runtime_path, -1, wruntime_path,
-                      sizeof(wruntime_path)/sizeof(wchar_t));
-  run_runtime(wruntime_path , cmdline);
-#if _MSC_VER >= 1200
-    __assume(0); /* Not reached */
-#endif
-#ifdef __MINGW32__
-    return 0;
-#endif
-}
diff --git a/tools/ocamlsize b/tools/ocamlsize
index 84798ce1b4dd0f55c401b7cf2435ccd91079c0fd..2ddb1959abdd7ef05301b9915887649fb35272b0 100755
--- a/tools/ocamlsize
+++ b/tools/ocamlsize
@@ -17,6 +17,19 @@
 
 foreach $f (@ARGV) {
     open(FILE, $f) || die("Cannot open $f");
+    read(FILE, $header, 2);
+    if ($header eq '#!') {
+      $path = <FILE>;
+      if ($path eq '/bin/sh') {
+        # exec form of the shebang header
+        $path = <FILE>;
+        if ($path =~ s/^exec '(.*)' "\$0" "\$@\"$/$1/ > 0) {
+          $path =~ s/'\\''/'/g;
+        } else {
+          undef $path;
+        }
+      }
+    };
     seek(FILE, -16, 2);
     $num_sections = &read_int();
     read(FILE, $magic, 12);
@@ -31,14 +44,16 @@ foreach $f (@ARGV) {
         $length{$sec} = $seclength[$i];
     }
     print $f, ":\n" if ($#ARGV > 0);
-    $path =
-        $length{'RNTM'} > 0 ?
-            &read_section('RNTM') :
-            "(default runtime)\n";
+    if (not defined $path) {
+      $path =
+          $length{'RNTM'} > 0 ?
+              substr(&read_section('RNTM'), 0, -1) :
+              "(custom runtime)";
+    };
     printf ("\tcode: %-7d data: %-7d symbols: %-7d debug: %-7d\n",
             $length{'CODE'}, $length{'DATA'},
             $length{'SYMB'}, $length{'DBUG'});
-    printf ("\tmagic number: %s  runtime system: %s",
+    printf ("\tmagic number: %s  runtime system: %s\n",
             $magic, $path);
     close(FILE);
 }
