diff --git a/configure b/configure
index 224f800..1beec50 100755
--- a/configure
+++ b/configure
@@ -4315,13 +4315,13 @@ if test "$ac_test_CFLAGS" = set; then
   CFLAGS=$ac_save_CFLAGS
 elif test $ac_cv_prog_cc_g = yes; then
   if test "$GCC" = yes; then
-    CFLAGS="-g -O2"
+    CFLAGS="-g3 -Og"
   else
-    CFLAGS="-g"
+    CFLAGS="-g3"
   fi
 else
   if test "$GCC" = yes; then
-    CFLAGS="-O2"
+    CFLAGS="-g3 -Og"
   else
     CFLAGS=
   fi
@@ -12791,7 +12791,7 @@ case $host in #(
     internal_cflags="-Wno-unused $cc_warnings \
 -fexcess-precision=standard"
         # TODO: see whether the code can be fixed to avoid -Wno-unused
-        common_cflags="-O2 -fno-strict-aliasing -fwrapv -mms-bitfields"
+        common_cflags="-g3 -Og -fno-strict-aliasing -fwrapv -mms-bitfields"
         internal_cppflags='-D__USE_MINGW_ANSI_STDIO=0 -DUNICODE -D_UNICODE'
         internal_cppflags="$internal_cppflags -DWINDOWS_UNICODE="
         internal_cppflags="${internal_cppflags}\$(WINDOWS_UNICODE)" ;; #(
@@ -12801,7 +12801,7 @@ esac ;; #(
   *) :
     case $ocaml_cv_cc_vendor in #(
   clang-*) :
-    common_cflags="-O2 -fno-strict-aliasing -fwrapv";
+    common_cflags="-g3 -Og -fno-strict-aliasing -fwrapv";
       internal_cflags="$cc_warnings -fno-common" ;; #(
   gcc-[012]-*) :
     # Some versions known to miscompile OCaml, e,g, 2.7.2.1, some 2.96.
@@ -12818,15 +12818,15 @@ $as_echo "$as_me: WARNING: Consider using GCC version 4.2 or above." >&2;};
       internal_cflags="$cc_warnings" ;; #(
   gcc-4-[234]) :
     # No -fexcess-precision option before GCC 4.5
-      common_cflags="-std=gnu99 -O2 -fno-strict-aliasing -fwrapv \
+      common_cflags="-std=gnu99 -g3 -Og -fno-strict-aliasing -fwrapv \
 -fno-builtin-memcmp";
       internal_cflags="$cc_warnings" ;; #(
   gcc-4-*) :
-    common_cflags="-std=gnu99 -O2 -fno-strict-aliasing -fwrapv \
+    common_cflags="-std=gnu99 -g3 -Og -fno-strict-aliasing -fwrapv \
 -fno-builtin-memcmp";
       internal_cflags="$cc_warnings -fexcess-precision=standard" ;; #(
   gcc-*) :
-    common_cflags="-O2 -fno-strict-aliasing -fwrapv";
+    common_cflags="-g3 -Og -fno-strict-aliasing -fwrapv";
       internal_cflags="$cc_warnings -fno-common \
 -fexcess-precision=standard" ;; #(
   msvc-*) :
