From a7caf1b6ce5d7f1a1961edb62eec3f58501794ae Mon Sep 17 00:00:00 2001
From: Jonah Beckford <9566106-jonahbeckford@users.noreply.gitlab.com>
Date: Mon, 25 May 2026 11:43:14 -0700
Subject: [PATCH] b04


diff --git a/toplevel/toploop.ml b/toplevel/toploop.ml
index 62a5b0023e8a2cb837e022bc42570a6e1fbd6f30..ac7f06a2138b08b55f9130e1cfca937707af6fbc 100644
--- a/toplevel/toploop.ml
+++ b/toplevel/toploop.ml
@@ -153,12 +153,12 @@ let find_ocamlinit () =
         let file = Filename.concat dir file in
         if Sys.file_exists file then Some file else None
   in
-  let home_dir () = getenv "HOME" in
+  let home_dir () = getenv (if Sys.win32 then "USERPROFILE" else "HOME") in
   let config_dir () =
-    if Sys.win32 then None else
-    match getenv "XDG_CONFIG_HOME" with
-    | Some _ as v -> v
-    | None ->
+    match (getenv "XDG_CONFIG_HOME", getenv "LOCALAPPDATA") with
+    | (Some _ as v), _ -> v
+    | None, (Some _ as v) when Sys.win32 -> v
+    | _ ->
         match home_dir () with
         | None -> None
         | Some dir -> Some (Filename.concat dir ".config")
-- 
2.53.0.windows.2

