From 6eb8df0ea28b2ba7d9f9ddbd91acc8065cd9b831 Mon Sep 17 00:00:00 2001
From: Jonah Beckford <9566106-jonahbeckford@users.noreply.gitlab.com>
Date: Mon, 25 May 2026 11:37:01 -0700
Subject: [PATCH] b05

diff --git a/configure b/configure
index bfa4b50..4017464 100644
--- a/configure
+++ b/configure
@@ -13088,7 +13088,11 @@ $as_echo "$as_me: WARNING: Consider using GCC version 4.2 or above." >&2;};
       internal_cflags="$cc_warnings -fno-common \
 -fexcess-precision=standard" ;; #(
   msvc-*) :
-    common_cflags="-nologo -O2 -Gy- -MD $cc_warnings"
+    if [ -n "${CFLAGS_MSVC}" ]; then
+      common_cflags="-nologo -Gy- ${CFLAGS_MSVC} $cc_warnings"
+    else
+      common_cflags="-nologo -O2 -Gy- -MD $cc_warnings"
+    fi
       common_cppflags="-D_CRT_SECURE_NO_DEPRECATE"
       internal_cppflags='-DUNICODE -D_UNICODE'
 
-- 
2.53.0.windows.2
