diff --git a/Makefile b/Makefile
index 70f10ea87f16a97e115a6c726b135e3f6475a23c..8001864f1302087e43a5935d1316951a8b4697e0 100644
--- a/Makefile
+++ b/Makefile
@@ -16,6 +16,8 @@
 # The main Makefile
 
 ROOTDIR = .
+SUBDIR_NAME =
+
 # NOTE: it is important that OCAMLLEX is defined *before* Makefile.common
 # gets included, so that its definition here takes precedence
 # over the one there.
@@ -75,8 +77,6 @@ PERVASIVES=$(STDLIB_MODULES) outcometree topdirs toploop
 
 LIBFILES=stdlib.cma std_exit.cmo *.cmi $(HEADER_NAME)
 
-COMPLIBDIR=$(LIBDIR)/compiler-libs
-
 TOPINCLUDES=$(addprefix -I otherlibs/,$(filter-out %threads,$(OTHERLIBRARIES)))
 ifeq "$(UNIX_OR_WIN32)" "unix"
 EXTRAPATH=
@@ -368,8 +368,6 @@ partialclean::
 	rm -f flexlink.opt$(EXE)
 endif # ifeq "$(BOOTSTRAPPING_FLEXDLL)" "false"
 
-INSTALL_COMPLIBDIR = $(DESTDIR)$(COMPLIBDIR)
-INSTALL_FLEXDLLDIR = $(INSTALL_LIBDIR)/flexdll
 FLEXDLL_MANIFEST = default$(filter-out _i386,_$(ARCH)).manifest
 
 DOC_FILES=\
@@ -378,71 +376,111 @@ DOC_FILES=\
   README.win32.adoc \
   LICENSE
 
+# COMPILER_ARTEFACT_DIRS adds the common compiler-libs directories as prefixes
+# to a sequence of patterns in the first argument, e.g.
+# $(call COMPILER_ARTEFACT_DIRS, *.cmi) expands to utils/*.cmi, parsing/*.cmi,
+# and so forth. Multiple wildcard patterns may be supplied. An optional second
+# argument includes additional directories beyond the common ones (e.g. asmcomp,
+# etc.)
+COMPILER_ARTEFACT_DIRS = \
+  $(foreach dir, \
+      utils parsing typing bytecomp file_formats lambda driver toplevel \
+      $(if $(filter-out undefined, $(origin 2)), $(2)), \
+    $(addprefix $(dir)/, $(1)))
+NATIVE_ARTEFACT_DIRS = \
+  asmcomp toplevel/native \
+  middle_end middle_end/closure middle_end/flambda middle_end/flambda/base_types
+
 # Installation
+# Historically, the install target dynamically installed what had been built,
+# for example, if only world had been built then make install simply didn't
+# install the native tools. That infrastructure is potentially convenient when
+# working on the compiler, but potentially masks bugs. It is better to have the
+# installation targets require everything configure mandated to have built.
+# There are three entry points to installation:
+#   install - installs everything
+#   installopt - installs the native code compiler _and_ the extra .opt tools
+#   installoptopt - intalls just the extra .opt tools
+# The installopt targets have been maintained for now, but may be removed in the
+# future.
 
-.PHONY: install
-install::
-	$(MKDIR) "$(INSTALL_BINDIR)"
-	$(MKDIR) "$(INSTALL_LIBDIR)"
-ifeq "$(SUPPORTS_SHARED_LIBRARIES)" "true"
-	$(MKDIR) "$(INSTALL_STUBLIBDIR)"
+ifeq "$(NATIVE_COMPILER)" "true"
+install: full-installoptopt
+	$(call INSTALL_END)
+else
+install: common-install
+	$(call INSTALL_END)
 endif
-	$(MKDIR) "$(INSTALL_COMPLIBDIR)"
-	$(MKDIR) "$(INSTALL_DOCDIR)"
+
+# These three targets are the slightly esoteric special sauce that avoid
+# recursive make invocations in the install targets.
+# There are three basic install recipies:
+# - The old install target is available to common-install, but never recurses to
+# - The old installopt target is available as both full-installopt and
+#   native-install
+# - The old installoptopt target is also available as full-installoptopt and
+#   installopt
+# These sets of recipies are then welded together by these three dependency
+# specifications
+# - When configured with --disable-native-compiler, the install target simply
+#   depends on common-install (see above)
+# - Otherwise, install depends on full-installoptopt (see above)
+# - The recipe for full-installoptopt installs the .opt versions of the tools,
+#   but it _depends on_ full-installopt.
+# - full-installopt installs the native compiler, but it _depends on_
+#   common-install
+installopt: native-install
+
+full-installopt: common-install
+
+full-installoptopt: full-installopt
+
+.PHONY: common-install
+common-install::
+	$(call INSTALL_BEGIN)
 	$(MAKE) -C runtime install
-	$(INSTALL_PROG) ocaml$(EXE) "$(INSTALL_BINDIR)"
+	$(call INSTALL_ITEMS, ocaml$(EXE), bin)
 ifeq "$(INSTALL_BYTECODE_PROGRAMS)" "true"
-	$(INSTALL_PROG) ocamlc$(EXE) "$(INSTALL_BINDIR)/ocamlc.byte$(EXE)"
+ifeq "$(NATIVE_COMPILER)" "true"
+	$(call INSTALL_ITEM, \
+	  ocamlc$(EXE), bin, , ocamlc.byte$(EXE))
+else
+	$(call INSTALL_ITEM, \
+	  ocamlc$(EXE), bin, , ocamlc.byte$(EXE), ocamlc$(EXE))
+endif
 endif
 	$(MAKE) -C stdlib install
 ifeq "$(INSTALL_BYTECODE_PROGRAMS)" "true"
-	$(INSTALL_PROG) lex/ocamllex$(EXE) \
-	  "$(INSTALL_BINDIR)/ocamllex.byte$(EXE)"
+ifeq "$(NATIVE_COMPILER)" "true"
+	$(call INSTALL_ITEM, \
+	  lex/ocamllex$(EXE), bin, , ocamllex.byte$(EXE))
+else
+	$(call INSTALL_ITEM, \
+	  lex/ocamllex$(EXE), bin, , ocamllex.byte$(EXE), ocamllex$(EXE))
+endif
 endif
-	$(INSTALL_PROG) yacc/ocamlyacc$(EXE) "$(INSTALL_BINDIR)"
-	$(INSTALL_DATA) \
-	   utils/*.cmi \
-	   parsing/*.cmi \
-	   typing/*.cmi \
-	   bytecomp/*.cmi \
-	   file_formats/*.cmi \
-	   lambda/*.cmi \
-	   driver/*.cmi \
-	   toplevel/*.cmi \
-	   "$(INSTALL_COMPLIBDIR)"
-	$(INSTALL_DATA) \
-	   toplevel/byte/*.cmi \
-	   "$(INSTALL_COMPLIBDIR)"
+	$(call INSTALL_ITEMS, yacc/ocamlyacc$(EXE), bin)
+	$(call INSTALL_ITEMS, \
+	    $(call COMPILER_ARTEFACT_DIRS, *.cmi), \
+	  lib, $(INSTALL_LIBDIR_COMPILERLIBS))
 ifeq "$(INSTALL_SOURCE_ARTIFACTS)" "true"
-	$(INSTALL_DATA) \
-	   utils/*.cmt utils/*.cmti utils/*.mli \
-	   parsing/*.cmt parsing/*.cmti parsing/*.mli \
-	   typing/*.cmt typing/*.cmti typing/*.mli \
-	   file_formats/*.cmt file_formats/*.cmti file_formats/*.mli \
-	   lambda/*.cmt lambda/*.cmti lambda/*.mli \
-	   bytecomp/*.cmt bytecomp/*.cmti bytecomp/*.mli \
-	   driver/*.cmt driver/*.cmti driver/*.mli \
-	   toplevel/*.cmt toplevel/*.cmti toplevel/*.mli \
-	   "$(INSTALL_COMPLIBDIR)"
-	$(INSTALL_DATA) \
-	   toplevel/byte/*.cmt \
-	   "$(INSTALL_COMPLIBDIR)"
+	$(call INSTALL_ITEMS, \
+	    $(call COMPILER_ARTEFACT_DIRS, *.cmt *.cmti *.mli), \
+	  lib, $(INSTALL_LIBDIR_COMPILERLIBS))
+	$(call INSTALL_ITEMS, toplevel/byte/*.cmt, \
+	  lib, $(INSTALL_LIBDIR_COMPILERLIBS))
 endif
-	$(INSTALL_DATA) \
-	  compilerlibs/*.cma \
-	  "$(INSTALL_COMPLIBDIR)"
-	$(INSTALL_DATA) \
-	   $(BYTESTART) $(TOPLEVELSTART) \
-	   "$(INSTALL_COMPLIBDIR)"
-	$(INSTALL_PROG) $(expunge) "$(INSTALL_LIBDIR)"
-	$(INSTALL_DATA) \
-	   toplevel/topdirs.cmi \
-	   "$(INSTALL_LIBDIR)"
+	$(call INSTALL_ITEMS, compilerlibs/*.cma, \
+	  lib, $(INSTALL_LIBDIR_COMPILERLIBS))
+	$(call INSTALL_ITEMS, \
+	    $(BYTESTART) $(TOPLEVELSTART), \
+	  lib, $(INSTALL_LIBDIR_COMPILERLIBS))
+	$(call INSTALL_ITEMS, $(expunge), libexec)
+	$(call INSTALL_ITEMS, toplevel/topdirs.cmi, lib)
 ifeq "$(INSTALL_SOURCE_ARTIFACTS)" "true"
-	$(INSTALL_DATA) \
-	   toplevel/topdirs.cmt \
-	   toplevel/topdirs.cmti toplevel/topdirs.mli \
-	   "$(INSTALL_LIBDIR)"
+	$(call INSTALL_ITEMS, \
+	    toplevel/topdirs.cmt toplevel/topdirs.cmti toplevel/topdirs.mli, \
+	  lib)
 endif
 	$(MAKE) -C tools install
 ifeq "$(UNIX_OR_WIN32)" "unix" # Install manual pages only on Unix
@@ -457,156 +495,109 @@ endif
 ifeq "$(WITH_OCAMLDOC)-$(STDLIB_MANPAGES)" "ocamldoc-true"
 	$(MAKE) -C api_docgen install
 endif
-	if test -n "$(WITH_DEBUGGER)"; then \
-	  $(MAKE) -C debugger install; \
-	fi
+ifneq "$(WITH_DEBUGGER)" ""
+	$(MAKE) -C debugger install
+endif
 ifeq "$(BOOTSTRAPPING_FLEXDLL)" "true"
 ifeq "$(TOOLCHAIN)" "msvc"
-	$(INSTALL_DATA) $(FLEXDLL_SOURCES)/$(FLEXDLL_MANIFEST) \
-    "$(INSTALL_BINDIR)/"
+	# Technically this should not be installed with "executable"
+	# permissions, but in practice that request will be ignored.
+	$(call INSTALL_ITEMS, $(FLEXDLL_SOURCES)/$(FLEXDLL_MANIFEST), bin)
 endif
 ifeq "$(INSTALL_BYTECODE_PROGRAMS)" "true"
-	$(INSTALL_PROG) \
-	  boot/flexlink.byte$(EXE) "$(INSTALL_BINDIR)/flexlink.byte$(EXE)"
-endif # ifeq "$(INSTALL_BYTECODE_PROGRAMS)" "true"
-	$(MKDIR) "$(INSTALL_FLEXDLLDIR)"
-	$(INSTALL_DATA) $(addprefix stdlib/flexdll/, $(FLEXDLL_OBJECTS)) \
-    "$(INSTALL_FLEXDLLDIR)"
-endif # ifeq "$(BOOTSTRAPPING_FLEXDLL)" "true"
-	$(INSTALL_DATA) Makefile.config "$(INSTALL_LIBDIR)"
-	$(INSTALL_DATA) $(DOC_FILES) "$(INSTALL_DOCDIR)"
-ifeq "$(INSTALL_BYTECODE_PROGRAMS)" "true"
-	if test -f ocamlopt$(EXE); then $(MAKE) installopt; else \
-	   cd "$(INSTALL_BINDIR)"; \
-	   $(LN) ocamlc.byte$(EXE) ocamlc$(EXE); \
-	   $(LN) ocamllex.byte$(EXE) ocamllex$(EXE); \
-	   (test -f flexlink.byte$(EXE) && \
-	      $(LN) flexlink.byte$(EXE) flexlink$(EXE)) || true; \
-	fi
+ifeq "$(NATIVE_COMPILER)" "true"
+	$(call INSTALL_ITEMS, boot/flexlink.byte$(EXE), bin)
 else
-	if test -f ocamlopt$(EXE); then $(MAKE) installopt; fi
+	$(call INSTALL_ITEM, boot/flexlink.byte$(EXE), bin, , , flexlink$(EXE))
 endif
+endif # ifeq "$(INSTALL_BYTECODE_PROGRAMS)" "true"
+	$(call INSTALL_ITEMS, \
+	    $(addprefix stdlib/flexdll/, $(FLEXDLL_OBJECTS)), \
+	  lib, $(INSTALL_LIBDIR_FLEXDLL))
+endif # ifeq "$(BOOTSTRAPPING_FLEXDLL)" "true"
+	$(call INSTALL_ITEMS, Makefile.config, lib)
+	$(call INSTALL_ITEMS, $(DOC_FILES), doc)
 
 # Installation of the native-code compiler
-.PHONY: installopt
-installopt:
+.PHONY: full-installopt native-install
+full-installopt native-install:
 	$(MAKE) -C runtime installopt
 ifeq "$(INSTALL_BYTECODE_PROGRAMS)" "true"
-	$(INSTALL_PROG) ocamlopt$(EXE) "$(INSTALL_BINDIR)/ocamlopt.byte$(EXE)"
+	$(call INSTALL_ITEM, ocamlopt$(EXE), bin, , ocamlopt.byte$(EXE))
 endif
 	$(MAKE) -C stdlib installopt
-	$(INSTALL_DATA) \
-	    middle_end/*.cmi \
-	    "$(INSTALL_COMPLIBDIR)"
-	$(INSTALL_DATA) \
-	    middle_end/closure/*.cmi \
-	    "$(INSTALL_COMPLIBDIR)"
-	$(INSTALL_DATA) \
-	    middle_end/flambda/*.cmi \
-	    "$(INSTALL_COMPLIBDIR)"
-	$(INSTALL_DATA) \
-	    middle_end/flambda/base_types/*.cmi \
-	    "$(INSTALL_COMPLIBDIR)"
-	$(INSTALL_DATA) \
-	    asmcomp/*.cmi \
-	    "$(INSTALL_COMPLIBDIR)"
+	$(call INSTALL_ITEMS, \
+	    middle_end/*.cmi, \
+	  lib, $(INSTALL_LIBDIR_COMPILERLIBS))
+	$(call INSTALL_ITEMS, \
+	    middle_end/closure/*.cmi, \
+	  lib, $(INSTALL_LIBDIR_COMPILERLIBS))
+	$(call INSTALL_ITEMS, \
+	    middle_end/flambda/*.cmi, \
+	  lib, $(INSTALL_LIBDIR_COMPILERLIBS))
+	$(call INSTALL_ITEMS, \
+	    middle_end/flambda/base_types/*.cmi, \
+	  lib, $(INSTALL_LIBDIR_COMPILERLIBS))
+	$(call INSTALL_ITEMS, \
+	    asmcomp/*.cmi, \
+	  lib, $(INSTALL_LIBDIR_COMPILERLIBS))
 ifeq "$(INSTALL_SOURCE_ARTIFACTS)" "true"
-	$(INSTALL_DATA) \
-	    middle_end/*.cmt middle_end/*.cmti \
-	    middle_end/*.mli \
-	    "$(INSTALL_COMPLIBDIR)"
-	$(INSTALL_DATA) \
-	    middle_end/closure/*.cmt middle_end/closure/*.cmti \
-	    middle_end/closure/*.mli \
-	    "$(INSTALL_COMPLIBDIR)"
-	$(INSTALL_DATA) \
-	    middle_end/flambda/*.cmt middle_end/flambda/*.cmti \
-	    middle_end/flambda/*.mli \
-	    "$(INSTALL_COMPLIBDIR)"
-	$(INSTALL_DATA) \
-	    middle_end/flambda/base_types/*.cmt \
-            middle_end/flambda/base_types/*.cmti \
-	    middle_end/flambda/base_types/*.mli \
-	    "$(INSTALL_COMPLIBDIR)"
-	$(INSTALL_DATA) \
-	    asmcomp/*.cmt asmcomp/*.cmti \
-	    asmcomp/*.mli \
-	    "$(INSTALL_COMPLIBDIR)"
+	$(call INSTALL_ITEMS, \
+	    $(addprefix middle_end/, *.cmt *.cmti *.mli), \
+	  lib, $(INSTALL_LIBDIR_COMPILERLIBS))
+	$(call INSTALL_ITEMS, \
+	    $(addprefix middle_end/closure/, *.cmt *.cmti *.mli), \
+	  lib, $(INSTALL_LIBDIR_COMPILERLIBS))
+	$(call INSTALL_ITEMS, \
+	    $(addprefix middle_end/flambda/, *.cmt *.cmti *.mli), \
+	  lib, $(INSTALL_LIBDIR_COMPILERLIBS))
+	$(call INSTALL_ITEMS, \
+	    $(addprefix middle_end/flambda/base_types/, *.cmt *.cmti *.mli), \
+	  lib, $(INSTALL_LIBDIR_COMPILERLIBS))
+	$(call INSTALL_ITEMS, \
+	    $(addprefix asmcomp/, *.cmt *.cmti *.mli), \
+	  lib, $(INSTALL_LIBDIR_COMPILERLIBS))
 endif
-	$(INSTALL_DATA) \
-	    $(OPTSTART) \
-	    "$(INSTALL_COMPLIBDIR)"
+	$(call INSTALL_ITEMS, $(OPTSTART), lib, $(INSTALL_LIBDIR_COMPILERLIBS))
 ifneq "$(WITH_OCAMLDOC)" ""
 	$(MAKE) -C ocamldoc installopt
 endif
 	for i in $(OTHERLIBRARIES); do \
 	  $(MAKE) -C otherlibs/$$i installopt || exit $$?; \
 	done
-ifeq "$(INSTALL_BYTECODE_PROGRAMS)" "true"
-	if test -f ocamlopt.opt$(EXE); then $(MAKE) installoptopt; else \
-	   cd "$(INSTALL_BINDIR)"; \
-	   $(LN) ocamlc.byte$(EXE) ocamlc$(EXE); \
-	   $(LN) ocamlopt.byte$(EXE) ocamlopt$(EXE); \
-	   $(LN) ocamllex.byte$(EXE) ocamllex$(EXE); \
-	   (test -f flexlink.byte$(EXE) && \
-	     $(LN) flexlink.byte$(EXE) flexlink$(EXE)) || true; \
-	fi
-else
-	if test -f ocamlopt.opt$(EXE); then $(MAKE) installoptopt; fi
-endif
 	$(MAKE) -C tools installopt
 
-.PHONY: installoptopt
-installoptopt:
-	$(INSTALL_PROG) ocamlc.opt$(EXE) "$(INSTALL_BINDIR)"
-	$(INSTALL_PROG) ocamlopt.opt$(EXE) "$(INSTALL_BINDIR)"
-	$(INSTALL_PROG) lex/ocamllex.opt$(EXE) "$(INSTALL_BINDIR)"
-	cd "$(INSTALL_BINDIR)"; \
-	   $(LN) ocamlc.opt$(EXE) ocamlc$(EXE); \
-	   $(LN) ocamlopt.opt$(EXE) ocamlopt$(EXE); \
-	   $(LN) ocamllex.opt$(EXE) ocamllex$(EXE)
+.PHONY: full-installoptopt installopt installoptopt
+full-installoptopt installopt installoptopt:
+	$(call INSTALL_ITEM, ocamlc.opt$(EXE), bin, , , ocamlc$(EXE))
+	$(call INSTALL_ITEM, ocamlopt.opt$(EXE), bin, , , ocamlopt$(EXE))
+	$(call INSTALL_ITEM, lex/ocamllex.opt$(EXE), bin, , , ocamllex$(EXE))
 ifeq "$(BOOTSTRAPPING_FLEXDLL)" "true"
-	$(INSTALL_PROG) flexlink.opt$(EXE) "$(INSTALL_BINDIR)"
-	cd "$(INSTALL_BINDIR)"; \
-	  $(LN) flexlink.opt$(EXE) flexlink$(EXE)
+	$(call INSTALL_ITEM, flexlink.opt$(EXE), bin, , , flexlink$(EXE))
 endif
-	$(INSTALL_DATA) \
-	   utils/*.cmx parsing/*.cmx typing/*.cmx bytecomp/*.cmx \
-	   toplevel/*.cmx toplevel/native/*.cmx \
-	   toplevel/native/tophooks.cmi \
-	   file_formats/*.cmx \
-	   lambda/*.cmx \
-	   driver/*.cmx asmcomp/*.cmx middle_end/*.cmx \
-           middle_end/closure/*.cmx \
-           middle_end/flambda/*.cmx \
-           middle_end/flambda/base_types/*.cmx \
-          "$(INSTALL_COMPLIBDIR)"
-	$(INSTALL_DATA) \
-	   compilerlibs/*.cmxa compilerlibs/*.$(A) \
-	   "$(INSTALL_COMPLIBDIR)"
-	$(INSTALL_DATA) \
-	   $(BYTESTART:.cmo=.cmx) $(BYTESTART:.cmo=.$(O)) \
-	   $(OPTSTART:.cmo=.cmx) $(OPTSTART:.cmo=.$(O)) \
-	   $(TOPLEVELSTART:.cmo=.$(O)) \
-	   "$(INSTALL_COMPLIBDIR)"
+	$(call INSTALL_ITEMS, \
+	    $(call COMPILER_ARTEFACT_DIRS, *.cmx, $(NATIVE_ARTEFACT_DIRS)) \
+	    toplevel/native/tophooks.cmi, \
+	  lib, $(INSTALL_LIBDIR_COMPILERLIBS))
+	$(call INSTALL_ITEMS, compilerlibs/*.cmxa compilerlibs/*.$(A), \
+	  lib, $(INSTALL_LIBDIR_COMPILERLIBS))
+	$(call INSTALL_ITEMS, \
+	    $(BYTESTART:.cmo=.$(O)) \
+	    $(OPTSTART:.cmo=.$(O)) \
+	    $(TOPLEVELSTART:.cmo=.$(O)), \
+	  lib, $(INSTALL_LIBDIR_COMPILERLIBS))
 ifeq "$(INSTALL_OCAMLNAT)" "true"
-	  $(INSTALL_PROG) ocamlnat$(EXE) "$(INSTALL_BINDIR)"
+	  $(call INSTALL_ITEMS, ocamlnat$(EXE), bin)
 endif
 
 # Installation of the *.ml sources of compiler-libs
 .PHONY: install-compiler-sources
 install-compiler-sources:
 ifeq "$(INSTALL_SOURCE_ARTIFACTS)" "true"
-	$(INSTALL_DATA) \
-	   utils/*.ml parsing/*.ml typing/*.ml bytecomp/*.ml driver/*.ml \
-           file_formats/*.ml \
-           lambda/*.ml \
-	   toplevel/*.ml toplevel/byte/*.ml \
-	   middle_end/*.ml middle_end/closure/*.ml \
-     middle_end/flambda/*.ml middle_end/flambda/base_types/*.ml \
-	   asmcomp/*.ml \
-	   asmcmp/debug/*.ml \
-	   "$(INSTALL_COMPLIBDIR)"
+	$(call INSTALL_ITEMS, \
+	    $(call COMPILER_ARTEFACT_DIRS, *.ml, $(NATIVE_ARTEFACT_DIRS)) \
+	    toplevel/byte/*.ml, \
+	  lib, $(INSTALL_LIBDIR_COMPILERLIBS))
 endif
 
 # Run all tests
diff --git a/Makefile.common b/Makefile.common
index 09db6b4b32615f1cc246240c144db8eeb6b0825c..7f07400aa0e0ee3885d4b29ac33739a5eea42665 100644
--- a/Makefile.common
+++ b/Makefile.common
@@ -41,13 +41,183 @@ endif
 QUOTE_SINGLE = '$(subst ','\'',$(1))'
 
 DESTDIR ?=
-INSTALL_BINDIR := $(DESTDIR)$(BINDIR)
-INSTALL_LIBDIR := $(DESTDIR)$(LIBDIR)
-INSTALL_STUBLIBDIR := $(DESTDIR)$(STUBLIBDIR)
-INSTALL_MANDIR := $(DESTDIR)$(MANDIR)
-INSTALL_PROGRAMS_MAN_DIR := $(DESTDIR)$(PROGRAMS_MAN_DIR)
-INSTALL_LIBRARIES_MAN_DIR := $(DESTDIR)$(LIBRARIES_MAN_DIR)
-INSTALL_DOCDIR := $(DESTDIR)$(DOCDIR)
+
+# Augment directories from Makefile.config / Makefile.build_config with
+# $(DESTDIR). i.e. each of these 5 directories may be overridden by the user,
+# and the compiler distribution makes no assumptions about where they are
+# relative to each other.
+INSTALL_BINDIR = $(DESTDIR)$(BINDIR)
+INSTALL_DOCDIR = $(DESTDIR)$(DOCDIR)
+INSTALL_LIBDIR = $(DESTDIR)$(LIBDIR)
+INSTALL_MANDIR = $(DESTDIR)$(MANDIR)
+INSTALL_STUBLIBDIR = $(DESTDIR)$(STUBLIBDIR)
+
+# Library subdirectories. The compiler distribution does make assumptions about
+# these, and they cannot be freely overridden by the user.
+INSTALL_LIBDIR_CAML = caml
+INSTALL_LIBDIR_COMPILERLIBS = compiler-libs
+INSTALL_LIBDIR_DYNLINK = dynlink
+INSTALL_LIBDIR_FLEXDLL = flexdll
+INSTALL_LIBDIR_OCAMLDOC = ocamldoc
+INSTALL_LIBDIR_SYSTHREADS = threads
+
+INSTALL_MANDIR_PROGRAMS = man1
+INSTALL_MANDIR_LIBRARIES = man3
+
+INSTALL_MODE ?= install
+
+# INSTALL_ITEM installs a single file, possibly with a different name and
+# possibly creating additional symlinks/copies
+#   $1 = source file (may include directories)
+#   $2 = section (bin, doc, lib, libexec, man, stublibs)
+#   $3 = directory within section (may be empty)
+#   $4 = target basename (may be empty)
+#   $5 = additional basenames (either symlinked or copied, depending on what the
+#        platform supports)
+# The $(origin n) dance is necessary to suppress warnings about undefined
+# variables.
+INSTALL_ITEM = \
+  $(INSTALL_$(INSTALL_MODE)_PREFIX)$(call INSTALL_ENSURE_DIR,$\
+    $(strip $(2)),$(if $(filter-out undefined,$(origin 3)),$(strip $(3))))$\
+  $(call INSTALL_DESPATCH_$(INSTALL_MODE)_ITEM,$\
+    $(strip $(1)),$\
+    $(strip $(2)),$\
+    $(if $(filter-out undefined,$(origin 3)),$(strip $(3))),$\
+    $(if $(filter-out undefined,$(origin 4)),$(strip $(4))),$\
+    $(if $(filter-out undefined,$(origin 5)),$(strip $(5))))
+
+# INSTALL_ITEMS installs a series of files to a single directory
+#   $1 = source file(s) (may include directories and glob patterns)
+#   $2 = section (as for INSTALL_ITEM)
+#   $3 = directory within section (may be omitted)
+# INSTALL_ITEMS is sometimes an alias for INSTALL_ITEM. For simplicity with
+# undefined variable warnings, INSTALL_DESPATCH_foo_ITEMS is passed 5 parameters
+# but $4 and $5 are always empty.
+INSTALL_ITEMS = \
+  $(INSTALL_$(INSTALL_MODE)_PREFIX)$(call INSTALL_ENSURE_DIR,$\
+    $(strip $(2)),$(if $(filter-out undefined,$(origin 3)),$(strip $(3))))$\
+  $(call INSTALL_DESPATCH_$(INSTALL_MODE)_ITEMS,$\
+    $(strip $(1)),$\
+    $(strip $(2)),$\
+    $(if $(filter-out undefined,$(origin 3)),$(strip $(3))),,)
+
+# INSTALL_ITEMS_OPT is INSTALL_ITEMS, but does nothing if the source file(s) do
+# not exist
+INSTALL_ITEMS_OPT = \
+  $(if $(wildcard $(1)),$(call INSTALL_ITEMS, \
+    $(1), $(2), $(if $(filter-out undefined,$(origin 3)), $(3))))
+
+INSTALL_ENSURE_DIR = \
+  $(if $(filter undefined,$(origin DIR_CREATED_$(subst exec,,$(1))_$(2))),$\
+    $(eval DIR_CREATED_$(1)_$(2):=)$\
+    $(call INSTALL_DESPATCH_$(INSTALL_MODE)_MKDIR,$\
+      $(subst exec,,$(1)),$(2)))
+
+# INSTALL_RM takes a single argument which may include glob patterns of files to
+# be removed when performing a physical install.
+INSTALL_RM = $(call INSTALL_DESPATCH_$(INSTALL_MODE)_RM,$(strip $(1)))
+
+# INSTALL_BEGIN and INSTALL_END are used in the root Makefile's install target
+INSTALL_BEGIN = $(INSTALL_DESPATCH_$(INSTALL_MODE)_BEGIN)
+INSTALL_END = $(INSTALL_DESPATCH_$(INSTALL_MODE)_END)
+
+# Normal installation
+INSTALL_CMD_bin = $(INSTALL_PROG)
+INSTALL_CMD_doc = $(INSTALL_DATA)
+INSTALL_CMD_lib = $(INSTALL_DATA)
+INSTALL_CMD_libexec = $(INSTALL_PROG)
+INSTALL_CMD_man = $(INSTALL_DATA)
+INSTALL_CMD_stublibs = $(INSTALL_PROG)
+
+INSTALL_SECTION_bin = $(INSTALL_BINDIR)
+INSTALL_SECTION_doc = $(INSTALL_DOCDIR)
+INSTALL_SECTION_lib = $(INSTALL_LIBDIR)
+INSTALL_SECTION_libexec = $(INSTALL_LIBDIR)
+INSTALL_SECTION_man = $(INSTALL_MANDIR)
+INSTALL_SECTION_stublibs = $(INSTALL_STUBLIBDIR)
+
+QUOTE_SINGLE = '$(subst ','\'',$(1))'
+
+define NEWLINE
+
+
+endef
+SH_AND = && \$(NEWLINE)
+
+INSTALL_install_PREFIX =
+
+INSTALL_DESPATCH_install_RM = rm -f $(1)
+
+INSTALL_DESPATCH_install_MKDIR = \
+  $(MKDIR) $(call QUOTE_SINGLE,$(INSTALL_SECTION_$(1))$(addprefix /,$(2))) \
+    $(SH_AND)
+
+MK_LINK = \
+  (cd "$(INSTALL_SECTION_$(2))$(addprefix /,$(3))" && \
+    $(LN) $(call QUOTE_SINGLE,$(1)) $(call QUOTE_SINGLE,$(4)))
+
+INSTALL_DESPATCH_install_ITEM = \
+  $(INSTALL_CMD_$(2)) $(1) \
+    $(call QUOTE_SINGLE,$\
+      $(INSTALL_SECTION_$(2))$(addprefix /,$(3))$(addprefix /,$(4))) \
+  $(foreach link, $(5),$(SH_AND)$\
+    $(call MK_LINK,$(if $(4),$(4),$(notdir $(1))),$(2),$(3),$(link)))
+
+INSTALL_DESPATCH_install_ITEMS = $(INSTALL_DESPATCH_install_ITEM)
+
+INSTALL_DESPATCH_install_BEGIN = @
+INSTALL_DESPATCH_install_END = @
+
+INSTALL_display_PREFIX = @
+
+INSTALL_DESPATCH_display_RM = @
+
+INSTALL_DESPATCH_display_MKDIR = \
+  echo $(call QUOTE_SINGLE,$\
+    -> MKDIR $(INSTALL_SECTION_$(1))$(addprefix /,$(2))) $(SH_AND)
+
+MKLINK_display = \
+  echo $(call QUOTE_SINGLE,-> LN \
+         $(abspath $(INSTALL_SECTION_$(2))$(addprefix /,$(3))/$(1)) -> \
+         $(if $(4),$(4),$(notdir $(1))))
+
+INSTALL_DESPATCH_display_ITEM = \
+  echo $(call QUOTE_SINGLE,-> INSTALL $(1) \
+         $(INSTALL_SECTION_$(2))$(addprefix /,$(3))$(addprefix /,$(4))) \
+  $(foreach link, $(5), && \
+    $(call MKLINK_display,$(if $(4),$(4),$(notdir $(1))),$(2),$(3),$(link)))
+
+INSTALL_DESPATCH_display_ITEMS = $(INSTALL_DESPATCH_display_ITEM)
+
+INSTALL_DESPATCH_display_BEGIN = @
+INSTALL_DESPATCH_display_END = @
+
+INSTALL_list_PREFIX = @
+
+INSTALL_DESPATCH_list_RM = @
+
+INSTALL_DESPATCH_list_MKDIR =
+
+MKLINK_list = \
+  echo $(call QUOTE_SINGLE,-> \
+         $(abspath $(INSTALL_SECTION_$(2))$(addprefix /,$(3))/$\
+                     $(if $(4),$(4),$(notdir $(1)))))
+
+INSTALL_DESPATCH_list_ITEM = \
+  echo $(call QUOTE_SINGLE,-> \
+         $(abspath $(INSTALL_SECTION_$(2))$(addprefix /,$(3))/$\
+                     $(if $(4),$(4),$(notdir $(1))))) \
+  $(foreach link, $(5), && \
+    $(call MKLINK_list,$(if $(4),$(4),$(notdir $(1))),$(2),$(3),$(link)))
+
+INSTALL_DESPATCH_list_ITEMS = \
+  $(foreach file, $(wildcard $(1)), \
+    echo $(call QUOTE_SINGLE,-> \
+           $(INSTALL_SECTION_$(2))$(addprefix /,$(3))/$(notdir $(file)));) \
+  true
+
+INSTALL_DESPATCH_list_BEGIN = @
+INSTALL_DESPATCH_list_END = @
 
 FLEXDLL_SUBMODULE_PRESENT := $(wildcard $(ROOTDIR)/flexdll/Makefile)
 
diff --git a/api_docgen/Makefile b/api_docgen/Makefile
index 07254645812ae43d8e1578e280debbd67e23c043..4ad898ea6bfd6802a4f83888ecdbfe0c983c429d 100644
--- a/api_docgen/Makefile
+++ b/api_docgen/Makefile
@@ -14,6 +14,7 @@
 #**************************************************************************
 # Used by included Makefiles
 ROOTDIR = ..
+SUBDIR_NAME = api_docgen
 -include ../Makefile.build_config
 
 odoc-%:
diff --git a/api_docgen/ocamldoc/Makefile b/api_docgen/ocamldoc/Makefile
index 6af89ab49e2c9ae0e07eb6346eec29a55b6b2f90..cf6c67166e1db724f877a500932c6bd311ef9c76 100644
--- a/api_docgen/ocamldoc/Makefile
+++ b/api_docgen/ocamldoc/Makefile
@@ -14,6 +14,7 @@
 #**************************************************************************
 # Used by included Makefiles
 ROOTDIR = ../..
+SUBDIR_NAME = api_docgen/ocamldoc
 include ../Makefile.common
 include ../../ocamldoc/Makefile.best_ocamldoc
 vpath %.mli ../../stdlib $(DOC_COMPILERLIBS_DIRS)  $(DOC_STDLIB_DIRS)
@@ -121,7 +122,4 @@ build/latex/compilerlibs_input.tex: | build/latex
 
 .PHONY: install
 install:
-	$(MKDIR) "$(INSTALL_LIBRARIES_MAN_DIR)"
-	if test -d build/man; then \
-	  $(INSTALL_DATA) build/man/*.3o "$(INSTALL_LIBRARIES_MAN_DIR)"; \
-	fi
+	$(call INSTALL_ITEMS_OPT, build/man/*.3o, man, $(INSTALL_MANDIR_LIBRARIES))
diff --git a/api_docgen/odoc/Makefile b/api_docgen/odoc/Makefile
index 5fc75b5cf18548565747b60ecc077413d610faa4..dfb4e9f7d8828a97c4c35b6d8ca1cc85f6344a6d 100644
--- a/api_docgen/odoc/Makefile
+++ b/api_docgen/odoc/Makefile
@@ -15,6 +15,7 @@
 
 # Used by included Makefiles
 ROOTDIR = ../..
+SUBDIR_NAME = api_docgen/odoc
 
 include ../Makefile.common
 
@@ -191,13 +192,10 @@ $(ALL_PAGED_DOC:%=build/%.3o.stamp):build/%.3o.stamp:build/%.odocl | build/
 # Man pages are the only installed documentation
 .PHONY: install
 install:
-	$(MKDIR) "$(INSTALL_LIBRARIES_MAN_DIR)"
-	if test -d build/man/libref ; then \
-	  $(INSTALL_DATA) build/man/libref/* "$(INSTALL_LIBRARIES_MAN_DIR)"; \
-	fi
-	if test -d build/man/compilerlibref ; then \
-	  $(INSTALL_DATA) build/man/libref/* "$(INSTALL_LIBRARIES_MAN_DIR)"; \
-	fi
+	$(call INSTALL_ITEMS_OPT, \
+	  build/man/libref/*, man, $(INSTALL_MANDIR_LIBRARIES))
+	$(call INSTALL_ITEMS_OPT, \
+	  build/man/compilerlibref/*, man, $(INSTALL_MANDIR_LIBRARIES))
 
 # Dependencies for stdlib modules.
 # Use the same dependencies used for compiling .cmx files.
diff --git a/debugger/Makefile b/debugger/Makefile
index 7e1ff1a74af7cdf14b51cc3214d82b6c0fe63951..08ab7d433dea742443880302a9f85d574c57d5b1 100644
--- a/debugger/Makefile
+++ b/debugger/Makefile
@@ -14,6 +14,7 @@
 #**************************************************************************
 
 ROOTDIR = ..
+SUBDIR_NAME = debugger
 
 include $(ROOTDIR)/Makefile.common
 include $(ROOTDIR)/Makefile.best_binaries
@@ -71,7 +72,7 @@ ocamldebug$(EXE): $(libraries) $(compiler_objects) ocamldebug.cmo \
 	$(CAMLC) $(LINKFLAGS) -o $@ -linkall $^
 
 install:
-	$(INSTALL_PROG) ocamldebug$(EXE) "$(INSTALL_BINDIR)"
+	$(call INSTALL_ITEMS, ocamldebug$(EXE), bin)
 
 clean::
 	rm -f ocamldebug ocamldebug.exe
diff --git a/man/Makefile b/man/Makefile
index 05424bab473741ca23b5e6899de7699e93a2470d..e43a3451bc66e19fa118fb4c70cc21c8badf1c2c 100644
--- a/man/Makefile
+++ b/man/Makefile
@@ -14,6 +14,7 @@
 #**************************************************************************
 
 ROOTDIR = ..
+SUBDIR_NAME = man
 include $(ROOTDIR)/Makefile.common
 
 MANPAGES = $(addsuffix .1,\
@@ -22,5 +23,4 @@ MANPAGES = $(addsuffix .1,\
 
 .PHONY: install
 install:
-	$(MKDIR) $(call QUOTE_SINGLE,$(INSTALL_PROGRAMS_MAN_DIR))
-	$(INSTALL_DATA) $(MANPAGES) $(call QUOTE_SINGLE,$(INSTALL_PROGRAMS_MAN_DIR))
+	$(call INSTALL_ITEMS, $(MANPAGES), man, $(INSTALL_MANDIR_PROGRAMS))
diff --git a/ocamldoc/Makefile b/ocamldoc/Makefile
index 74cc1e05f54a2dca7e065ef6ec2581008969becc..2a4b3813db74b94cd33a0a86938bc1381490de1a 100644
--- a/ocamldoc/Makefile
+++ b/ocamldoc/Makefile
@@ -14,6 +14,7 @@
 #**************************************************************************
 
 ROOTDIR = ..
+SUBDIR_NAME = ocamldoc
 
 include $(ROOTDIR)/Makefile.common
 include $(ROOTDIR)/Makefile.best_binaries
@@ -239,19 +240,12 @@ DEPEND_PREREQS = $(LEXERS:.mll=.ml) \
 
 .PHONY: install
 install:
-	$(MKDIR) "$(INSTALL_BINDIR)"
-	$(MKDIR) "$(INSTALL_LIBDIR)/ocamldoc"
-	$(INSTALL_PROG) $(OCAMLDOC) "$(INSTALL_BINDIR)"
-	$(INSTALL_DATA) \
-	  ocamldoc.hva *.cmi $(OCAMLDOC_LIBCMA) \
-	  "$(INSTALL_LIBDIR)/ocamldoc"
-	$(INSTALL_DATA) \
-	  $(OCAMLDOC_LIBCMIS) \
-	  "$(INSTALL_LIBDIR)/ocamldoc"
+	$(call INSTALL_ITEMS, ocamldoc$(EXE), bin)
+	$(call INSTALL_ITEMS, ocamldoc.hva *.cmi $(OCAMLDOC_LIBCMA), \
+	  lib, $(INSTALL_LIBDIR_OCAMLDOC))
 ifeq "$(INSTALL_SOURCE_ARTIFACTS)" "true"
-	$(INSTALL_DATA) \
-	  $(OCAMLDOC_LIBMLIS) $(OCAMLDOC_LIBCMTS) \
-	  "$(INSTALL_LIBDIR)/ocamldoc"
+	$(call INSTALL_ITEMS, $(OCAMLDOC_LIBMLIS) $(OCAMLDOC_LIBCMTS), \
+	  lib, $(INSTALL_LIBDIR_OCAMLDOC))
 endif
 
 # Note: at the moment, $(INSTALL_MANODIR) is created even if the doc has
@@ -263,20 +257,10 @@ installopt:
 
 .PHONY: installopt_really
 installopt_really:
-	$(MKDIR) "$(INSTALL_BINDIR)"
-	$(MKDIR) "$(INSTALL_LIBDIR)/ocamldoc"
-	$(INSTALL_PROG) $(OCAMLDOC_OPT) "$(INSTALL_BINDIR)"
-	$(INSTALL_DATA) \
-	  $(OCAMLDOC_LIBCMIS) \
-	  "$(INSTALL_LIBDIR)/ocamldoc"
-ifeq "$(INSTALL_SOURCE_ARTIFACTS)" "true"
-	$(INSTALL_DATA) \
-	  $(OCAMLDOC_LIBMLIS) $(OCAMLDOC_LIBCMTS) \
-	  "$(INSTALL_LIBDIR)/ocamldoc"
-endif
-	$(INSTALL_DATA) \
-	  ocamldoc.hva *.cmx $(OCAMLDOC_LIBA) $(OCAMLDOC_LIBCMXA) \
-	  "$(INSTALL_LIBDIR)/ocamldoc"
+	$(call INSTALL_ITEMS, ocamldoc.opt$(EXE), bin)
+	$(call INSTALL_ITEMS, \
+	  *.cmx $(OCAMLDOC_LIBA) $(OCAMLDOC_LIBCMXA), \
+	  lib, $(INSTALL_LIBDIR_OCAMLDOC))
 
 # TODO: also split into several rules
 
diff --git a/otherlibs/Makefile b/otherlibs/Makefile
index 0da6669b731f3bb7c329a3b84a30bba59409c5d4..dbe9c62a0916fa39f8062648bf4265e0b975a0ec 100644
--- a/otherlibs/Makefile
+++ b/otherlibs/Makefile
@@ -14,6 +14,7 @@
 #**************************************************************************
 
 ROOTDIR=..
+SUBDIR_NAME=otherlibs
 include $(ROOTDIR)/Makefile.common
 
 OTHERLIBRARIES ?= bigarray dynlink str systhreads \
diff --git a/otherlibs/Makefile.otherlibs.common b/otherlibs/Makefile.otherlibs.common
index e730336a4235c2a018569ffd9a01befaed1e2b67..69e7e5dba9880bd64a6c3060ac276c169b827ba6 100644
--- a/otherlibs/Makefile.otherlibs.common
+++ b/otherlibs/Makefile.otherlibs.common
@@ -16,6 +16,7 @@
 # Common Makefile for otherlibs
 
 ROOTDIR=../..
+SUBDIR_NAME?=otherlibs/$(LIBNAME)
 include $(ROOTDIR)/Makefile.common
 include $(ROOTDIR)/Makefile.best_binaries
 
@@ -104,31 +105,24 @@ lib$(CLIBNAME).$(A): $(COBJS)
 install::
 ifneq "$(STUBSLIB)" ""
 ifeq "$(SUPPORTS_SHARED_LIBRARIES)" "true"
-	$(INSTALL_PROG) $(STUBSDLL) "$(INSTALL_STUBLIBDIR)"
+	$(call INSTALL_ITEMS, $(STUBSDLL), stublibs)
 endif
-	$(INSTALL_DATA) $(STUBSLIB) "$(INSTALL_LIBDIR)/"
+	$(call INSTALL_ITEMS, $(STUBSLIB), lib)
 endif
-
-	$(INSTALL_DATA) \
-	  $(LIBNAME).cma $(CMIFILES) \
-	  "$(INSTALL_LIBDIR)/"
+	$(call INSTALL_ITEMS, $(LIBNAME).cma $(CMIFILES), lib)
 ifeq "$(INSTALL_SOURCE_ARTIFACTS)" "true"
-	$(INSTALL_DATA) \
-	  $(CMIFILES:.cmi=.mli) \
-          $(CMIFILES:.cmi=.cmti) \
-	  "$(INSTALL_LIBDIR)/"
+	$(call INSTALL_ITEMS, $(CMIFILES:.cmi=.mli) $(CMIFILES:.cmi=.cmti), lib)
+endif
+ifneq "$(HEADERS)" ""
+	$(call INSTALL_ITEMS, $(HEADERS), lib, $(INSTALL_LIBDIR_CAML))
 endif
-	if test -n "$(HEADERS)"; then \
-	  $(INSTALL_DATA) $(HEADERS) "$(INSTALL_LIBDIR)/caml/"; \
-	fi
 
 installopt:
-	$(INSTALL_DATA) \
-	   $(CAMLOBJS_NAT) $(LIBNAME).cmxa $(LIBNAME).$(A) \
-	   "$(INSTALL_LIBDIR)/"
-	if test -f $(LIBNAME).cmxs; then \
-	  $(INSTALL_PROG) $(LIBNAME).cmxs "$(INSTALL_LIBDIR)"; \
-	fi
+	$(call INSTALL_ITEMS, \
+	   $(CAMLOBJS_NAT) $(LIBNAME).cmxa $(LIBNAME).$(A), lib)
+ifeq "$(SUPPORTS_SHARED_LIBRARIES)" "true"
+	$(call INSTALL_ITEMS, $(LIBNAME).cmxs, libexec)
+endif
 
 partialclean:
 	rm -f *.cm*
diff --git a/otherlibs/dynlink/Makefile b/otherlibs/dynlink/Makefile
index 6526e8bf3980b29a2648260fc4d812bea0dadc89..f9984daee0bb83ae2da278916ac0930526587c2b 100644
--- a/otherlibs/dynlink/Makefile
+++ b/otherlibs/dynlink/Makefile
@@ -20,6 +20,7 @@
 # FIXME reduce redundancy by including ../Makefile
 
 ROOTDIR = ../..
+SUBDIR_NAME = otherlibs/dynlink
 
 include $(ROOTDIR)/Makefile.common
 include $(ROOTDIR)/Makefile.best_binaries
@@ -242,20 +243,14 @@ $(extract_crc): dynlink.cma byte/dynlink_compilerlibs.cmo extract_crc.cmo
 	$(OCAMLC) -o $@ $^
 
 install:
-	$(INSTALL_DATA) \
-	  dynlink.cmi dynlink.cma \
-	  "$(INSTALL_LIBDIR)"
+	$(call INSTALL_ITEMS, dynlink.cmi dynlink.cma, lib)
 ifeq "$(INSTALL_SOURCE_ARTIFACTS)" "true"
-	$(INSTALL_DATA) \
-	  dynlink.cmti dynlink.mli \
-	  "$(INSTALL_LIBDIR)"
+	$(call INSTALL_ITEMS, dynlink.cmti dynlink.mli, lib)
 endif
 
 installopt:
 ifeq "$(strip $(NATDYNLINK))" "true"
-	$(INSTALL_DATA) \
-	  $(NATOBJS) dynlink.cmxa dynlink.$(A) \
-	  "$(INSTALL_LIBDIR)"
+	$(call INSTALL_ITEMS, $(NATOBJS) dynlink.cmxa dynlink.$(A), lib)
 endif
 
 partialclean:
diff --git a/otherlibs/systhreads/Makefile b/otherlibs/systhreads/Makefile
index b22b645f45700806baf13beb85b3fcb1434006b0..72d517bb94b3f74a344dbb8ccec132886342a974 100644
--- a/otherlibs/systhreads/Makefile
+++ b/otherlibs/systhreads/Makefile
@@ -14,6 +14,7 @@
 #**************************************************************************
 
 ROOTDIR=../..
+SUBDIR_NAME=otherlibs/systhreads
 
 include $(ROOTDIR)/Makefile.common
 include $(ROOTDIR)/Makefile.best_binaries
@@ -111,30 +112,22 @@ clean: partialclean
 	rm -f dllthreads*.so dllthreads*.dll *.a *.lib *.o *.obj
 	rm -rf $(DEPDIR)
 
-INSTALL_THREADSLIBDIR=$(INSTALL_LIBDIR)/$(LIBNAME)
-
 install:
 ifeq "$(SUPPORTS_SHARED_LIBRARIES)" "true"
-	$(INSTALL_PROG) $(DLLTHREADS) "$(INSTALL_STUBLIBDIR)"
+	$(call INSTALL_ITEMS, $(DLLTHREADS), stublibs)
 endif
-	$(INSTALL_DATA) libthreads.$(A) "$(INSTALL_LIBDIR)"
-	mkdir -p "$(INSTALL_THREADSLIBDIR)"
-	$(INSTALL_DATA) \
-	  $(CMIFILES) threads.cma \
-	  "$(INSTALL_THREADSLIBDIR)"
+	$(call INSTALL_ITEMS, libthreads.$(A), lib)
+	$(call INSTALL_ITEMS, $(CMIFILES) threads.cma, lib, $(LIBNAME))
 ifeq "$(INSTALL_SOURCE_ARTIFACTS)" "true"
-	$(INSTALL_DATA) \
-	  $(CMIFILES:.cmi=.cmti) \
-	  "$(INSTALL_THREADSLIBDIR)"
-	$(INSTALL_DATA) $(MLIFILES) "$(INSTALL_THREADSLIBDIR)"
+	$(call INSTALL_ITEMS, $(CMIFILES:.cmi=.cmti), lib, $(LIBNAME))
+	$(call INSTALL_ITEMS, $(MLIFILES), lib, $(LIBNAME))
 endif
-	$(INSTALL_DATA) threads.h "$(INSTALL_LIBDIR)/caml"
+	$(call INSTALL_ITEMS, threads.h, lib, $(INSTALL_LIBDIR_CAML))
 
 installopt:
-	$(INSTALL_DATA) libthreadsnat.$(A) "$(INSTALL_LIBDIR)"
-	$(INSTALL_DATA) \
-	  $(THREADS_NCOBJS) threads.cmxa threads.$(A) \
-	  "$(INSTALL_THREADSLIBDIR)"
+	$(call INSTALL_ITEMS, libthreadsnat.$(A), lib)
+	$(call INSTALL_ITEMS, $(THREADS_NCOBJS) threads.cmxa threads.$(A), \
+	  lib, $(LIBNAME))
 
 %.cmi: %.mli
 	$(CAMLC) -c $(COMPFLAGS) $<
diff --git a/otherlibs/win32unix/Makefile b/otherlibs/win32unix/Makefile
index 35371fe7d0bcf2b8efed8d974287dd9e15ed43ae..be9df075a233f74b14d7fdecdcfca0818f458c52 100644
--- a/otherlibs/win32unix/Makefile
+++ b/otherlibs/win32unix/Makefile
@@ -13,6 +13,8 @@
 #*                                                                        *
 #**************************************************************************
 
+SUBDIR_NAME = otherlibs/win32unix
+
 # Note: since this directory is Windows-specific, it may be good to make sure
 # its content can not be compiled under Unix.
 # This directory could even become a subdirectory of the unix directory.
diff --git a/runtime/Makefile b/runtime/Makefile
index e5cfb24282aeb84be5c44844c80f0190501888ad..7f907ab709c9b267f21cfdbfe3b52a4de8f47785 100644
--- a/runtime/Makefile
+++ b/runtime/Makefile
@@ -14,6 +14,7 @@
 #**************************************************************************
 
 ROOTDIR = ..
+SUBDIR_NAME = runtime
 
 include $(ROOTDIR)/Makefile.common
 
@@ -158,7 +159,6 @@ allopt:
 	$(error The build has been configured with --disable-native-compiler)
 endif
 
-INSTALL_INCDIR=$(INSTALL_LIBDIR)/caml
 .PHONY: install
 
 ifeq "$(SUFFIXING)" "true"
@@ -171,15 +171,9 @@ endif
 
 define INSTALL_RUNTIME
 install::
-	$(INSTALL_PROG) \
-    $(1)$(EXE) "$(INSTALL_BINDIR)/$(call MANGLE_RUNTIME_NAME,$(1))"
-ifeq "$(SUFFIXING)" "true"
-	cd "$(INSTALL_BINDIR)" && \
-    $(LN) "$(TARGET)-$(1)-$(BYTECODE_RUNTIME_ID)$(EXE)" "$(1)$(EXE)"
-	cd "$(INSTALL_BINDIR)" && \
-    $(LN) "$(TARGET)-$(1)-$(BYTECODE_RUNTIME_ID)$(EXE)" \
-	    "$(1)-$(ZINC_RUNTIME_ID)$(EXE)"
-endif
+	$$(call INSTALL_ITEM, $(1)$(EXE), bin, , \
+    $(call MANGLE_RUNTIME_NAME,$(1)), $(if $(filter true, $(SUFFIXING)), \
+      $(1)$(EXE) $(1)-$(ZINC_RUNTIME_ID)$(EXE)))
 endef
 define INSTALL_RUNTIME_LIB
 ifeq "$(2)" "BYTECODE"
@@ -187,32 +181,27 @@ install::
 else
 installopt::
 endif
-	$(INSTALL_PROG) \
-    lib$(1)_shared$(EXT_DLL) \
-	    "$(INSTALL_LIBDIR)/$(call MANGLE_RUNTIME_DLL_NAME,$(1),$(2))"
-ifeq "$(SUFFIXING)" "true"
-	cd "$(INSTALL_LIBDIR)" && \
-    $(LN) "$(call MANGLE_RUNTIME_DLL_NAME,$(1),$(2))" \
-    "lib$(1)_shared$(EXT_DLL)"
-endif
+	$$(call INSTALL_ITEM, lib$(1)_shared$(EXT_DLL), libexec, , \
+	   $(call MANGLE_RUNTIME_DLL_NAME,$(1),$(2)), \
+	     $(if $(filter true, $(SUFFIXING)), lib$(1)_shared$(EXT_DLL)))
 endef
 
 $(foreach runtime, $(PROGRAMS), \
   $(eval $(call INSTALL_RUNTIME,$(runtime))))
 
 install::
-	$(INSTALL_DATA) ld.conf $(BYTECODE_STATIC_LIBRARIES) "$(INSTALL_LIBDIR)"
+	$(call INSTALL_ITEMS, ld.conf $(BYTECODE_STATIC_LIBRARIES), lib)
 
 $(foreach shared_runtime, $(BYTECODE_SHARED_LIBRARIES), \
   $(eval $(call INSTALL_RUNTIME_LIB,$(shared_runtime),BYTECODE)))
 
 install::
-	mkdir -p "$(INSTALL_INCDIR)"
-	$(INSTALL_DATA) caml/domain_state.tbl caml/*.h "$(INSTALL_INCDIR)"
+	$(call INSTALL_ITEMS, caml/domain_state.tbl caml/*.h, \
+	  lib, $(INSTALL_LIBDIR_CAML))
 
 .PHONY: installopt
 installopt::
-	$(INSTALL_DATA) $(NATIVE_STATIC_LIBRARIES) "$(INSTALL_LIBDIR)"
+	$(call INSTALL_ITEMS, $(NATIVE_STATIC_LIBRARIES), lib)
 
 $(foreach shared_runtime, $(NATIVE_SHARED_LIBRARIES), \
   $(eval $(call INSTALL_RUNTIME_LIB,$(shared_runtime),NATIVE)))
diff --git a/stdlib/Makefile b/stdlib/Makefile
index 41b5f10c89f959e17d1b129d50973defe3d5362e..d0e410722afcbbba574370736eae3ed07c220746 100644
--- a/stdlib/Makefile
+++ b/stdlib/Makefile
@@ -14,6 +14,7 @@
 #**************************************************************************
 
 ROOTDIR = ..
+SUBDIR_NAME = stdlib
 
 include $(ROOTDIR)/Makefile.common
 
@@ -80,25 +81,26 @@ install::
     exit 1; \
   fi
 
+# bigarray.mli is only installed if the legacy bigarray library is _not_ being
+# compiled (since its bigarray.mli will override ours)
 install::
-	$(INSTALL_DATA) \
-	  stdlib.cma std_exit.cmo *.cmi camlheader_ur \
-	  "$(INSTALL_LIBDIR)"
+	$(call INSTALL_ITEMS, stdlib.cma std_exit.cmo *.cmi camlheader_ur, lib)
 ifeq "$(INSTALL_SOURCE_ARTIFACTS)" "true"
-	$(INSTALL_DATA) \
-	  *.cmt *.cmti *.mli *.ml \
-	  "$(INSTALL_LIBDIR)"
+	$(call INSTALL_ITEMS, \
+	  *.cmt *.cmti *.ml \
+	  $(if $(filter bigarray, $(OTHERLIBRARIES)), \
+	    $(filter-out bigarray.mli, $(wildcard *.mli)), *.mli), lib)
 endif
-	$(INSTALL_DATA) $(HEADER_NAME) "$(INSTALL_LIBDIR)/$(HEADER_NAME)"
+	$(call INSTALL_ITEMS, $(HEADER_NAME), lib)
 
 ifeq "$(RUNTIMED)" "true"
 install::
-	$(INSTALL_DATA) target_camlheaderd "$(INSTALL_LIBDIR)/camlheaderd"
+	$(call INSTALL_ITEM, target_camlheaderd, lib, , camlheaderd)
 endif
 
 ifeq "$(INSTRUMENTED_RUNTIME)" "true"
 install::
-	$(INSTALL_DATA) target_camlheaderi "$(INSTALL_LIBDIR)/camlheaderi"
+	$(call INSTALL_ITEM, target_camlheaderi, lib, , camlheaderi)
 endif
 
 .PHONY: installopt
@@ -116,9 +118,7 @@ installopt-default::
   fi
 
 installopt-default::
-	$(INSTALL_DATA) \
-	  stdlib.cmxa stdlib.$(A) std_exit.$(O) *.cmx \
-	  "$(INSTALL_LIBDIR)"
+	$(call INSTALL_ITEMS, stdlib.cmxa stdlib.$(A) std_exit.$(O) *.cmx, lib)
 
 ifeq "$(SHEBANGSCRIPTS)" "true"
 
diff --git a/tools/Makefile b/tools/Makefile
index 5ff2dd99f2a3f15ab466e89f225b1551905a92e5..ac120fc3fe53408af158f1853fe03e2803a608fe 100644
--- a/tools/Makefile
+++ b/tools/Makefile
@@ -14,6 +14,7 @@
 #**************************************************************************
 
 ROOTDIR = ..
+SUBDIR_NAME = tools
 # NOTE: it is important that OCAMLLEX is defined *before* Makefile.common
 # gets included, so that its definition here takes precedence
 # over the one there.
@@ -110,19 +111,13 @@ ocamloptp.opt$(EXE): $(call byte2native, $(OCAMLCP) ocamloptp.cmo)
 opt:: profiling.cmx
 
 install::
-	$(INSTALL_DATA) \
-	  profiling.cmi profiling.cmo \
-	  "$(INSTALL_LIBDIR)"
+	$(call INSTALL_ITEMS, profiling.cmi profiling.cmo, lib)
 ifeq "$(INSTALL_SOURCE_ARTIFACTS)" "true"
-	$(INSTALL_DATA) \
-	  profiling.cmt profiling.cmti \
-	  "$(INSTALL_LIBDIR)"
+	$(call INSTALL_ITEMS, profiling.cmt profiling.cmti, lib)
 endif
 
 installopt::
-	$(INSTALL_DATA) \
-          profiling.cmx profiling.$(O) \
-	  "$(INSTALL_LIBDIR)"
+	$(call INSTALL_ITEMS, profiling.cmx profiling.$(O), lib)
 
 # To help building mixed-mode libraries (OCaml + C)
 OCAMLMKLIB = config.cmo build_path_prefix_map.cmo misc.cmo ocamlmklib.cmo
@@ -143,27 +138,18 @@ ocamlmktop.opt$(EXE): $(call byte2native, $(OCAMLMKTOP))
 
 LIBRARY3=config.cmo build_path_prefix_map.cmo misc.cmo warnings.cmo location.cmo
 
+define INSTALL_ONE_TOOL
 install::
 ifeq "$(INSTALL_BYTECODE_PROGRAMS)" "true"
-	for i in $(install_files); \
-	do \
-	  $(INSTALL_PROG) "$$i$(EXE)" "$(INSTALL_BINDIR)/$$i.byte$(EXE)"; \
-	  if test -f "$$i".opt$(EXE); then \
-	    $(INSTALL_PROG) "$$i.opt$(EXE)" "$(INSTALL_BINDIR)" && \
-	    (cd "$(INSTALL_BINDIR)" && $(LN) "$$i.opt$(EXE)" "$$i$(EXE)"); \
-	  else \
-	    (cd "$(INSTALL_BINDIR)" && $(LN) "$$i.byte$(EXE)" "$$i$(EXE)"); \
-	  fi; \
-	done
-else
-	for i in $(install_files); \
-	do \
-	  if test -f "$$i".opt$(EXE); then \
-	    $(INSTALL_PROG) "$$i.opt$(EXE)" "$(INSTALL_BINDIR)"; \
-	    (cd "$(INSTALL_BINDIR)" && $(LN) "$$i.opt$(EXE)" "$$i$(EXE)"); \
-	  fi; \
-	done
+	$(call INSTALL_ITEM, $(1)$(EXE), bin, , $(1).byte$(EXE), \
+	                     $(if $(wildcard $(1).opt$(EXE)),, $(1)$(EXE)))
 endif
+	$(if $(wildcard $(1).opt$(EXE)), \
+	  $(call INSTALL_ITEM, $(1).opt$(EXE), bin, , , $(1)$(EXE)))
+endef
+
+$(foreach tool, $(install_files), \
+  $(eval $(call INSTALL_ONE_TOOL,$(tool))))
 
 # The preprocessor for asm generators
 
@@ -190,12 +176,9 @@ ocamlcmt$(EXE): $(OCAMLCMT)
 ocamlcmt.opt$(EXE): $(call byte2native, $(OCAMLCMT))
 
 install::
-	if test -f ocamlcmt.opt$(EXE); then \
-	  $(INSTALL_PROG)\
-	    ocamlcmt.opt$(EXE) "$(INSTALL_BINDIR)/ocamlcmt$(EXE)"; \
-	else \
-	  $(INSTALL_PROG) ocamlcmt$(EXE) "$(INSTALL_BINDIR)"; \
-	fi
+	$(call INSTALL_ITEM, \
+	  ocamlcmt$(if $(wildcard ocamlcmt.opt$(EXE)),.opt)$(EXE), bin, , \
+	  ocamlcmt$(EXE))
 
 # The bytecode disassembler
 
@@ -266,9 +249,7 @@ clean::
 # Eventlog metadata file
 
 install::
-	$(INSTALL_DATA) \
-	  eventlog_metadata \
-	  "$(INSTALL_LIBDIR)"
+	$(call INSTALL_ITEMS, eventlog_metadata, lib)
 
 # Copy a bytecode executable, stripping debug info
 
