diff --git a/runtime/bigarray.c b/runtime/bigarray.c
--- a/runtime/bigarray.c
+++ b/runtime/bigarray.c
@@ -179,8 +179,8 @@ CAMLexport int caml_ba_compare(value v1, value v2)
   int i;
 
   /* Compare kind & layout in case the arguments are of different types */
-  flags1 = b1->flags & (CAML_BA_KIND_MASK | CAML_BA_LAYOUT_MASK);
-  flags2 = b2->flags & (CAML_BA_KIND_MASK | CAML_BA_LAYOUT_MASK);
+  flags1 = b1->flags & ((intnat) CAML_BA_KIND_MASK | (intnat) CAML_BA_LAYOUT_MASK);
+  flags2 = b2->flags & ((intnat) CAML_BA_KIND_MASK | (intnat) CAML_BA_LAYOUT_MASK);
   if (flags1 != flags2) return flags2 - flags1;
   /* Compare number of dimensions */
   if (b1->num_dims != b2->num_dims) return b2->num_dims - b1->num_dims;
@@ -373,7 +373,7 @@ CAMLexport void caml_ba_serialize(value v,
 
   /* Serialize header information */
   caml_serialize_int_4(b->num_dims);
-  caml_serialize_int_4(b->flags & (CAML_BA_KIND_MASK | CAML_BA_LAYOUT_MASK));
+  caml_serialize_int_4(b->flags & ((intnat) CAML_BA_KIND_MASK | (intnat) CAML_BA_LAYOUT_MASK));
   for (i = 0; i < b->num_dims; i++) {
     intnat len = b->dim[i];
     if (len < 0xffff) {
@@ -516,7 +516,7 @@ CAMLprim value caml_ba_create(value vkind, value vlayout, value vdim)
     if (dim[i] < 0)
       caml_invalid_argument("Bigarray.create: negative dimension");
   }
-  flags = Caml_ba_kind_val(vkind) | Caml_ba_layout_val(vlayout);
+  flags = Caml_ba_kind_val(vkind) | (intnat) Caml_ba_layout_val(vlayout);
   return caml_ba_alloc(flags, num_dims, NULL, dim);
 }
 
@@ -997,8 +997,8 @@ CAMLprim value caml_ba_change_layout(value vb, value vlayout)
   /* if the layout is different, change the flags and reverse the dimensions */
   if (Caml_ba_layout_val(vlayout) != (b->flags & CAML_BA_LAYOUT_MASK)) {
     /* change the flags to reflect the new layout */
-    int flags = (b->flags & (CAML_BA_KIND_MASK | CAML_BA_MANAGED_MASK))
-                 | Caml_ba_layout_val(vlayout);
+    int flags = (b->flags & ((intnat) CAML_BA_KIND_MASK | (intnat) CAML_BA_MANAGED_MASK))
+                 | (intnat) Caml_ba_layout_val(vlayout);
     /* reverse the dimensions */
     intnat new_dim[CAML_BA_MAX_NUM_DIMS];
     unsigned int i;
