diff --git a/runtime/caml/misc.h b/runtime/caml/misc.h
index c605f87..41c5e50 100644
--- a/runtime/caml/misc.h
+++ b/runtime/caml/misc.h
@@ -114,7 +114,7 @@ CAMLdeprecated_typedef(addr, char *);
 
 #ifndef CAMLDLLIMPORT
   #if defined(SUPPORT_DYNAMIC_LINKING) && defined(ARCH_SIXTYFOUR) \
-      && (defined(__CYGWIN__) || defined(__MINGW32__))
+      && (defined(__CYGWIN__) || defined(__MINGW32__) || defined(_MSC_VER))
     #define CAMLDLLIMPORT __declspec(dllimport)
   #else
     #define CAMLDLLIMPORT
